
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../modules/">
      
      
        <link rel="next" href="../usage_stats/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>MiXCR - Repseq Library for Immune repertoire postanalysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#usage-working-with-mixcr" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-header__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Repseq Library for Immune repertoire postanalysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              MiXCR
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../installation/" class="md-tabs__link">
        
  
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../modules/" class="md-tabs__link">
        
  
  
    
  
  Modules

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../functions/" class="md-tabs__link">
        
  
  
    
  
  Functions

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-nav__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Repseq Library for Immune repertoire postanalysis
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Modules
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Usage
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    MiXCR
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    MiXCR
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#working-with-metadata-and-creating-a-dataframe-with-samples" class="md-nav__link">
    <span class="md-ellipsis">
      
        Working with metadata and creating a dataframe with samples
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-template-for-mixcr-analyze" class="md-nav__link">
    <span class="md-ellipsis">
      
        Command template for mixcr analyze
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-mixcr-analyze-in-batches-using-slurm" class="md-nav__link">
    <span class="md-ellipsis">
      
        Running mixcr analyze in batches using SLURM
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#making-report-images" class="md-nav__link">
    <span class="md-ellipsis">
      
        Making report images
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#creating-a-table-containing-clonosets-stats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Creating a table containing clonosets stats
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stats
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_intersections/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intersections
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_clustering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Clustering
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_diffexp/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Differential expression
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Functions
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              
                



  


  <nav class="md-path" aria-label="Navigation" >
    <ol class="md-path__list">
      
        
  
  
    <li class="md-path__item">
      <a href=".." class="md-path__link">
        
  <span class="md-ellipsis">
    Home
  </span>

      </a>
    </li>
  

      
      
        
  
  
    
    
      <li class="md-path__item">
        <a href="./" class="md-path__link">
          
  <span class="md-ellipsis">
    Usage
  </span>

        </a>
      </li>
    
  

      
    </ol>
  </nav>

              
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="usage-working-with-mixcr">Usage: working with MiXCR</h1>
<p>MiXCR is the leading software for generating clonoset tables from raw FastQ files. <a href="../functions/#mixcr">MiXCR module</a> allows to run MiXCR 4.3+ batch analyses with SLURM queue manager.</p>
<div class="admonition note">
<p class="admonition-title">Setting up the environment</p>
<p>Before getting started, make sure that main_repseq environment is chosen. Otherwise, check the <a href="../installation/">installation guide</a></p>
</div>
<h2 id="working-with-metadata-and-creating-a-dataframe-with-samples">Working with metadata and creating a dataframe with samples</h2>
<p>Create <code>sample_df</code> from dataset metadata in <code>.yaml</code> format (if it's in a tabular format, use external libraries such as Pandas). Remove unnesessary columns if needed. Note If METADATA_FILENAME is absent, it is set to <code>metadata.yaml</code> by default. Note that this is relevant only for metadata created with NGSiK in CDR3.net group. 
<br>If your dataset does not have metadata, create the dataframe manually. <b>The neccessary columns are: <code>R1</code>, <code>R2</code>, <code>sample_id</code></b>, where <code>R1</code> and <code>R2</code> contain paths (using full paths is strongly advised) to respective raw files, and sample_id are arbitrary unique identificators.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">mixcr</span> <span class="k">as</span> <span class="n">mx</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">slurm</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">io</span> <span class="k">as</span> <span class="n">repseqio</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="n">sample_df</span> <span class="o">=</span> <span class="n">repseqio</span><span class="o">.</span><span class="n">read_yaml_metadata</span><span class="p">(</span><span class="n">RAW_DATA_DIR</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="n">METADATA_FILENAME</span><span class="p">)</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="n">metadata</span> <span class="o">=</span> <span class="n">sample_df</span><span class="o">.</span><span class="n">prop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">,</span> <span class="s1">&#39;R2&#39;</span><span class="p">])</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="n">output_dir</span> <span class="o">=</span> <span class="o">...</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="n">path_to_mixcr_binary</span> <span class="o">=</span> <span class="o">...</span>
</code></pre></div>
<p><code>sample_df</code> example:</p>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">R1</th>
<th style="text-align: left;">R2</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">sample_1_nCD4</td>
<td style="text-align: left;">/home/user/samples/sample1_nCD4_1_TRB_L001_R1_001.fastq.gz</td>
<td style="text-align: left;">/home/user/samples/sample1_nCD4_1_TRB_L001_R2_001.fastq.gz</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">sample_2_nCD4</td>
<td style="text-align: left;">/home/user/samples/sample2_nCD4_1_TRB_L001_R1_001.fastq.gz</td>
<td style="text-align: left;">/home/user/samples/sample2_nCD4_1_TRB_L001_R2_001.fastq.gz</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">sample_3_nCD4</td>
<td style="text-align: left;">/home/user/samples/sample3_nCD4_1_TRB_L001_R1_001.fastq.gz</td>
<td style="text-align: left;">/home/user/samples/sample3_nCD4_1_TRB_L001_R2_001.fastq.gz</td>
</tr>
</tbody>
</table>
<p><br></p>
<h2 id="command-template-for-mixcr-analyze">Command template for <code>mixcr analyze</code></h2>
<p>Create a command template for mixcr analyze. The default template is <code>mixcr analyze milab-human-rna-tcr-umi-multiplex -f r1 r2 output_prefix</code> for Milab Hum TCR RNA multiplex kit. The default values are 32 GB for <code>memory</code> (required OOM in GB),  1.5 hours for <code>time_estimate</code> and 40 for <code>cpus</code> (in case of Aldan3 server, it is the size of a smallest node). Note that <code>mixcr analyze</code> and <code>r1 r2 output_prefix</code> are "magical" parts of the template that should be kept as-is in the template, so change only the part in-between these parts.
<br>For more detailed information on MiXCR presets, visit the <a href="https://mixcr.com/mixcr/reference/overview-built-in-presets/">MiXCR website</a>.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="n">mixcr_race_command_template</span> <span class="o">=</span> <span class="s2">&quot;mixcr analyze milab-mouse-rna-tcr-umi-race -f r1 r2 output_prefix&quot;</span>
</code></pre></div>
<br></p>
<h2 id="running-mixcr-analyze-in-batches-using-slurm">Running <code>mixcr analyze</code> in batches using SLURM</h2>
<p>Run mixcr analyze in batches (Relevant only for servers using <b>SLURM</b>). This function generates a set of commands for each sample by creating a SLURM script for each command and submitting them to the SLURM queue. Scripts itself and .log files (contain std.out and std.error outputs) will be saved in <code>~/temp/SLURM</code>. </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="n">mx</span><span class="o">.</span><span class="n">mixcr4_analyze_batch</span><span class="p">(</span><span class="n">sample_df</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">command_template</span><span class="o">=</span><span class="n">mixcr_race_command_template</span><span class="p">,</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>                        <span class="n">path_to_mixcr_binary</span><span class="p">)</span>
</code></pre></div>
<p><br></p>
<p>To check the progress, use <code>check_slurm_progress</code>. <code>loop</code> set to <code>True</code> gives real-time updates with 0.5 s interval while <code>loop</code>=<code>False</code> shows current progress and runs in the background without blocking other cells. 
<br>Currently, <code>check_slurm_progress</code> might be unreliable in some cases, thus it's recommended to check SLURM queue and .log files in <code>~/temp/SLURM</code> folder manually.</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="n">slurm</span><span class="o">.</span><span class="n">check_slurm_progress</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;mixcr_analyze_slurm_batch.log&quot;</span><span class="p">),</span> <span class="n">loop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<br></p>
<h2 id="making-report-images">Making report images</h2>
<p>Make reports (combines <code>mixcr exportQc align</code>, <code>chainUsage</code> and <code>tags</code>) and get report images (both .pdf and .svg for <code>align</code> and <code>chainUsage</code>, only .pdf for <code>tags</code>). To see report images examples, visit the <a href="https://mixcr.com/mixcr/reference/qc-overview/">MiXCR website</a>.</p>
<ul>
<li>align — exports various quality control metrics</li>
<li>chainUsage — calculates chain usage across all clonotypes</li>
<li>tags — for samples with barcodes, provides barcode coverage statistics for every sample</li>
</ul>
<p>To see progress, use <code>check_slurm_progress</code> as shown below</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="n">mx</span><span class="o">.</span><span class="n">mixcr4_reports</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">mixcr_path</span><span class="o">=</span><span class="n">path_to_mixcr_binary</span><span class="p">)</span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="n">slurm</span><span class="o">.</span><span class="n">check_slurm_progress</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;mixcr_reports_slurm_batch.log&quot;</span><span class="p">),</span> <span class="n">loop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="n">mx</span><span class="o">.</span><span class="n">show_report_images</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
</code></pre></div>
<p><img alt="alignQc" src="../images_docs/alignQc.svg" /></p>
<p><img alt="chainsQc" src="../images_docs/chainsQc.svg" /></p>
<p><br></p>
<h2 id="creating-a-table-containing-clonosets-stats">Creating a table containing clonosets stats</h2>
<p>Get a tabular report using <code>get_processing_table</code> function. It searches for clonosets in the the folder, extracts their sample_id's and shows main
processing stats in a table format. By default does not show "off-target" clonosets - 
those having less than 1% (default, may be overriden) of reads for the sample_id.
For example, you have sequenced TRB sample, but there is found 0.5% (by read count) 
of TRA chains for the same sample_id, then the clonoset will not be shown in the table.
You can specify <code>show_offtarget=True</code> to display all found chains in the table or 
outherwise set a higher value for <code>offtarget_chain_threshold</code> (<code>0.01</code> by default).</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="n">proc_table</span> <span class="o">=</span> <span class="n">mx</span><span class="o">.</span><span class="n">get_processing_table</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>
</code></pre></div>
<p>A full processing table example:</p>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">extracted_chain</th>
<th style="text-align: right;">reads_total</th>
<th style="text-align: right;">reads_with_umi_pc</th>
<th style="text-align: right;">reads_aligned_pc</th>
<th style="text-align: right;">reads_overlapped_aln_pc</th>
<th style="text-align: right;">total_umi</th>
<th style="text-align: right;">umi_after_correction</th>
<th style="text-align: right;">overseq_threshold</th>
<th style="text-align: right;">reads_after_filter</th>
<th style="text-align: right;">umi_after_filter</th>
<th style="text-align: right;">reads_per_umi</th>
<th style="text-align: right;">clones_total</th>
<th style="text-align: right;">reads_in_clones_total</th>
<th style="text-align: right;">clones</th>
<th style="text-align: right;">reads_in_clones</th>
<th style="text-align: right;">clones_func</th>
<th style="text-align: right;">reads_in_func_clones</th>
<th style="text-align: right;">umi_in_clones</th>
<th style="text-align: right;">umi_in_func_clones</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">sample1_nCD4_1_TRB</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: right;">2120957</td>
<td style="text-align: right;">98.63</td>
<td style="text-align: right;">86.38</td>
<td style="text-align: right;">4.97</td>
<td style="text-align: right;">597401</td>
<td style="text-align: right;">564176</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1612478</td>
<td style="text-align: right;">344972</td>
<td style="text-align: right;">4.67</td>
<td style="text-align: right;">145019</td>
<td style="text-align: right;">1566962</td>
<td style="text-align: right;">145012</td>
<td style="text-align: right;">1566949</td>
<td style="text-align: right;">135644</td>
<td style="text-align: right;">1509856</td>
<td style="text-align: right;">349587</td>
<td style="text-align: right;">337223</td>
</tr>
<tr>
<td style="text-align: right;">3</td>
<td style="text-align: left;">sample2_nCD4_1_TRB</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: right;">1484339</td>
<td style="text-align: right;">98.64</td>
<td style="text-align: right;">85.43</td>
<td style="text-align: right;">3.49</td>
<td style="text-align: right;">242845</td>
<td style="text-align: right;">223565</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1183345</td>
<td style="text-align: right;">155008</td>
<td style="text-align: right;">7.63</td>
<td style="text-align: right;">69771</td>
<td style="text-align: right;">1167512</td>
<td style="text-align: right;">69770</td>
<td style="text-align: right;">1167509</td>
<td style="text-align: right;">66237</td>
<td style="text-align: right;">1133350</td>
<td style="text-align: right;">156334</td>
<td style="text-align: right;">151913</td>
</tr>
<tr>
<td style="text-align: right;">6</td>
<td style="text-align: left;">sample3_nCD4_1_TRB</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: right;">940861</td>
<td style="text-align: right;">96.49</td>
<td style="text-align: right;">86.42</td>
<td style="text-align: right;">2.84</td>
<td style="text-align: right;">279362</td>
<td style="text-align: right;">236455</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">734134</td>
<td style="text-align: right;">157646</td>
<td style="text-align: right;">4.66</td>
<td style="text-align: right;">68466</td>
<td style="text-align: right;">715176</td>
<td style="text-align: right;">68465</td>
<td style="text-align: right;">715174</td>
<td style="text-align: right;">64208</td>
<td style="text-align: right;">690533</td>
<td style="text-align: right;">160301</td>
<td style="text-align: right;">154995</td>
</tr>
</tbody>
</table>
<p>Some columns may be omitted for better readability:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="n">small_proc_table</span> <span class="o">=</span> <span class="n">proc_table</span><span class="p">[[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;extracted_chain&quot;</span><span class="p">,</span><span class="s2">&quot;reads_total&quot;</span><span class="p">,</span> <span class="s2">&quot;reads_with_umi_pc&quot;</span><span class="p">,</span> <span class="s2">&quot;reads_aligned_pc&quot;</span><span class="p">,</span> <span class="s2">&quot;reads_per_umi&quot;</span><span class="p">,</span> <span class="s2">&quot;overseq_threshold&quot;</span><span class="p">,</span><span class="s2">&quot;clones_func&quot;</span><span class="p">,</span> <span class="s2">&quot;umi_in_func_clones&quot;</span><span class="p">]]</span>
</code></pre></div>
<p>Columns:</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Parameters</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">sample_id</td>
<td style="text-align: left;">sample_id specified in the clonoset filename</td>
</tr>
<tr>
<td style="text-align: left;">extracted_chain</td>
<td style="text-align: left;">clonoset сhain specified in the clonoset filename (e.g., TRA, IGH). One sample (particular <code>sample_id</code>) might include multiple chains depending on the protocol used and the values of the <code>show_offtarget</code> and the <code>offtarget_chain_threshold</code> parameters</td>
</tr>
<tr>
<td style="text-align: left;">reads_total</td>
<td style="text-align: left;">a total number of raw reads for the whole sample</td>
</tr>
<tr>
<td style="text-align: left;">reads_with_umi_pc</td>
<td style="text-align: left;">Percentage of reads with barcodes if tag-pattern was used</td>
</tr>
<tr>
<td style="text-align: left;">reads_aligned_pc</td>
<td style="text-align: left;">Percentage of successfully aligned reads for the whole sample</td>
</tr>
<tr>
<td style="text-align: left;">reads_overlapped_aln_pc</td>
<td style="text-align: left;">Percentage of overlapping reads of all aligned reads</td>
</tr>
<tr>
<td style="text-align: left;">total_umi</td>
<td style="text-align: left;">Total number of UMIs before any filtering</td>
</tr>
<tr>
<td style="text-align: left;">umi_after_correction</td>
<td style="text-align: left;">Number of UMIs after PCR and sequencing error correction</td>
</tr>
<tr>
<td style="text-align: left;">overseq_threshold</td>
<td style="text-align: left;">Reads per group threshold used for filtering and selected by MiXCR</td>
</tr>
<tr>
<td style="text-align: left;">reads_after_filter</td>
<td style="text-align: left;">Number of reads after filtering</td>
</tr>
<tr>
<td style="text-align: left;">umi_after_filter</td>
<td style="text-align: left;">Number of UMIs after filtering</td>
</tr>
<tr>
<td style="text-align: left;">reads_per_umi</td>
<td style="text-align: left;">Average number of reads per UMI</td>
</tr>
<tr>
<td style="text-align: left;">clones_total</td>
<td style="text-align: left;">Total number of clonotypes according to MiXCR assemble report</td>
</tr>
<tr>
<td style="text-align: left;">reads_in_clones_total</td>
<td style="text-align: left;">Number of reads assigned to clonotypes for the whole sample</td>
</tr>
<tr>
<td style="text-align: left;">clones</td>
<td style="text-align: left;">Number of clonotypes in a clonoset for a given chain</td>
</tr>
<tr>
<td style="text-align: left;">reads_in_clones</td>
<td style="text-align: left;">Number of reads in a clonoset for a given chain</td>
</tr>
<tr>
<td style="text-align: left;">clones_func</td>
<td style="text-align: left;">Number of functional clonotypes (no frameshifts and stops)</td>
</tr>
<tr>
<td style="text-align: left;">reads_in_func_clones</td>
<td style="text-align: left;">Number of reads in functional clonotypes</td>
</tr>
<tr>
<td style="text-align: left;">umi_in_clones</td>
<td style="text-align: left;">Number of UMIs in a clonoset for a given chain</td>
</tr>
<tr>
<td style="text-align: left;">umi_in_func_clones</td>
<td style="text-align: left;">Number of UMIs in functional clonotypes</td>
</tr>
</tbody>
</table>
<p><br></p>
<details class="info">
<summary>Visualization</summary>
<p>Properties from proc_table can be visualized in Jupyter notebook using %%R cell magic. 
<img alt="proc_table" src="../images_docs/proc_table.png" /></p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="o">%</span><span class="n">load_ext</span> <span class="n">rpy2</span><span class="o">.</span><span class="n">ipython</span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="o">%%</span><span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">proc_table</span> <span class="o">-</span><span class="n">w</span> <span class="mi">900</span> <span class="o">-</span><span class="n">h</span> <span class="mi">500</span>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="n">params_order</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;reads_with_umi_pc&quot;</span><span class="p">,</span> <span class="s2">&quot;reads_aligned_pc&quot;</span><span class="p">,</span><span class="s2">&quot;reads_per_umi&quot;</span><span class="p">,</span><span class="s2">&quot;umi_in_func_clones&quot;</span><span class="p">,</span><span class="s2">&quot;clones_func&quot;</span><span class="p">)</span>
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a>
<a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="n">proc_table</span> <span class="o">%&gt;%</span>
<a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a>    <span class="n">select</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">experimental_group</span><span class="p">,</span> <span class="n">subset</span><span class="p">,</span> <span class="n">reads_per_umi</span><span class="p">,</span> <span class="n">reads_with_umi_pc</span><span class="p">,</span> <span class="n">reads_aligned_pc</span><span class="p">,</span> <span class="n">clones_func</span><span class="p">,</span> <span class="n">umi_in_func_clones</span><span class="p">)</span> <span class="o">%&gt;%</span>
<a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a>    <span class="n">pivot_longer</span><span class="p">(</span><span class="o">-</span><span class="n">c</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">experimental_group</span><span class="p">,</span> <span class="n">subset</span><span class="p">),</span> <span class="n">names_to</span><span class="o">=</span><span class="s2">&quot;parameter&quot;</span><span class="p">,</span> <span class="n">values_to</span><span class="o">=</span><span class="s2">&quot;value&quot;</span><span class="p">)</span> <span class="o">%&gt;%</span>
<a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a>    <span class="n">mutate</span><span class="p">(</span><span class="n">experimental_group</span><span class="o">=</span><span class="n">factor</span><span class="p">(</span><span class="n">experimental_group</span><span class="p">,</span> <span class="n">group_order</span><span class="p">))</span> <span class="o">%&gt;%</span>
<a id="__codelineno-7-10" name="__codelineno-7-10" href="#__codelineno-7-10"></a>    <span class="n">mutate</span><span class="p">(</span><span class="n">parameter</span><span class="o">=</span><span class="n">factor</span><span class="p">(</span><span class="n">parameter</span><span class="p">,</span> <span class="n">params_order</span><span class="p">))</span> <span class="o">%&gt;%</span>
<a id="__codelineno-7-11" name="__codelineno-7-11" href="#__codelineno-7-11"></a>    <span class="n">ggplot</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">experimental_group</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">value</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">experimental_group</span><span class="p">))</span> <span class="o">+</span>
<a id="__codelineno-7-12" name="__codelineno-7-12" href="#__codelineno-7-12"></a>        <span class="n">geom_boxplot</span><span class="p">(</span><span class="n">outlier</span><span class="o">.</span><span class="n">shape</span><span class="o">=</span><span class="n">NA</span><span class="p">)</span><span class="o">+</span>
<a id="__codelineno-7-13" name="__codelineno-7-13" href="#__codelineno-7-13"></a>        <span class="n">geom_jitter</span><span class="p">()</span><span class="o">+</span>
<a id="__codelineno-7-14" name="__codelineno-7-14" href="#__codelineno-7-14"></a>        <span class="n">facet_wrap</span><span class="p">(</span><span class="nb">vars</span><span class="p">(</span><span class="n">parameter</span><span class="p">),</span> <span class="n">scales</span><span class="o">=</span><span class="s2">&quot;free_y&quot;</span><span class="p">)</span><span class="o">+</span>
<a id="__codelineno-7-15" name="__codelineno-7-15" href="#__codelineno-7-15"></a>        <span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">colors_6_groups</span><span class="p">)</span> <span class="o">+</span> 
<a id="__codelineno-7-16" name="__codelineno-7-16" href="#__codelineno-7-16"></a>        <span class="n">boxplot_theme</span><span class="o">+</span>
<a id="__codelineno-7-17" name="__codelineno-7-17" href="#__codelineno-7-17"></a>        <span class="n">theme</span><span class="p">(</span><span class="n">legend</span><span class="o">.</span><span class="n">position</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">)</span>
</code></pre></div>
</details>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2023 <a href="https://github.com/mmjmike" target="_blank" rel="noopener">Mikhail Myshkin</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/mmjmike" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M202.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M496 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.path", "toc.integrate", "navigation.top", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>