
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../usage_clustering/">
      
      
        <link rel="next" href="../functions/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>Differential expression - Repseq Library for Immune repertoire postanalysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#usage-differential-expression-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-header__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Repseq Library for Immune repertoire postanalysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Differential expression
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../installation/" class="md-tabs__link">
        
  
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../modules/" class="md-tabs__link">
        
  
  
    
  
  Modules

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../usage_mixcr/" class="md-tabs__link">
          
  
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../functions/" class="md-tabs__link">
        
  
  
    
  
  Functions

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-nav__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Repseq Library for Immune repertoire postanalysis
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Modules
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Usage
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_mixcr/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MiXCR
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stats
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_intersections/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intersections
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_clustering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Clustering
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Differential expression
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Differential expression
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Input data
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#count-table-by-cluster" class="md-nav__link">
    <span class="md-ellipsis">
      
        Count table by cluster
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#differential-expression-calculation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Differential expression calculation
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#merge-cluster-properties-by-feature_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        Merge cluster properties by feature_id
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Functions
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              
                



  


  <nav class="md-path" aria-label="Navigation" >
    <ol class="md-path__list">
      
        
  
  
    <li class="md-path__item">
      <a href=".." class="md-path__link">
        
  <span class="md-ellipsis">
    Home
  </span>

      </a>
    </li>
  

      
      
        
  
  
    
    
      <li class="md-path__item">
        <a href="../usage_mixcr/" class="md-path__link">
          
  <span class="md-ellipsis">
    Usage
  </span>

        </a>
      </li>
    
  

      
    </ol>
  </nav>

              
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="usage-differential-expression-analysis">Usage: differential expression analysis</h1>
<div class="admonition note">
<p class="admonition-title">Setting up the environment</p>
<p>Before getting started, make sure that main_repseq environment is chosen. Otherwise, check the <a href="../installation/">installation guide</a></p>
</div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="n">REPSEQ_PATH</span> <span class="o">=</span> <span class="s1">&#39;your_path_to_repseq&#39;</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">REPSEQ_PATH</span><span class="p">)</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">io</span> <span class="k">as</span> <span class="n">repseqio</span>
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">common_functions</span> <span class="k">as</span> <span class="n">cf</span>
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">mixcr</span> <span class="k">as</span> <span class="n">mx</span>
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">slurm</span>
<a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">clonosets</span> <span class="k">as</span> <span class="n">cl</span>
<a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
<a id="__codelineno-0-10" name="__codelineno-0-10" href="#__codelineno-0-10"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">clone_filter</span> <span class="k">as</span> <span class="n">clf</span>
<a id="__codelineno-0-11" name="__codelineno-0-11" href="#__codelineno-0-11"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">intersections</span>
<a id="__codelineno-0-12" name="__codelineno-0-12" href="#__codelineno-0-12"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">clustering</span>
<a id="__codelineno-0-13" name="__codelineno-0-13" href="#__codelineno-0-13"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">diffexp</span>
<a id="__codelineno-0-14" name="__codelineno-0-14" href="#__codelineno-0-14"></a><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<a id="__codelineno-0-15" name="__codelineno-0-15" href="#__codelineno-0-15"></a>
<a id="__codelineno-0-16" name="__codelineno-0-16" href="#__codelineno-0-16"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<a id="__codelineno-0-17" name="__codelineno-0-17" href="#__codelineno-0-17"></a><span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<a id="__codelineno-0-18" name="__codelineno-0-18" href="#__codelineno-0-18"></a><span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<a id="__codelineno-0-19" name="__codelineno-0-19" href="#__codelineno-0-19"></a><span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
</code></pre></div>
<h2 id="input-data">Input data</h2>
<p>To run differential expression analysis, two dataframes are required:</p>
<ul>
<li><code>metadata</code> dataframe. Note that <b>sample_id</b> (must be unique and match <code>count_table</code> column names) and <b>group</b> columns are <b>nessesary</b> for this pipeline. Example:</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">Subset</th>
<th style="text-align: right;">Replica</th>
<th style="text-align: right;">Number_of_cells</th>
<th style="text-align: left;">group</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">Luk_TRB_AdV_CD4_1</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1376</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">Luk_TRB_AdV_CD4_2</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1504</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">Luk_TRB_AdV_CD4_3</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1623</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">3</td>
<td style="text-align: left;">Luk_TRB_AdV_CD8_1</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">37698</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">4</td>
<td style="text-align: left;">Luk_TRB_AdV_CD8_2</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">21359</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">5</td>
<td style="text-align: left;">Luk_TRB_AdV_CD8_3</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">21031</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">6</td>
<td style="text-align: left;">Luk_TRB_CMV_CD4_1</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1018</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">7</td>
<td style="text-align: left;">Luk_TRB_CMV_CD4_2</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">1032</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">8</td>
<td style="text-align: left;">Luk_TRB_CMV_CD4_3</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">2788</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">9</td>
<td style="text-align: left;">Luk_TRB_CMV_CD8_1</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">19617</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">10</td>
<td style="text-align: left;">Luk_TRB_CMV_CD8_2</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">17490</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">11</td>
<td style="text-align: left;">Luk_TRB_CMV_CD8_3</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">16662</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">12</td>
<td style="text-align: left;">Luk_TRB_EBV_CD4_1</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">453</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">13</td>
<td style="text-align: left;">Luk_TRB_EBV_CD4_2</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">840</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">14</td>
<td style="text-align: left;">Luk_TRB_EBV_CD4_3</td>
<td style="text-align: left;">CD4</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">770</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">15</td>
<td style="text-align: left;">Luk_TRB_EBV_CD8_1</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">41274</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">16</td>
<td style="text-align: left;">Luk_TRB_EBV_CD8_2</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">22020</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">17</td>
<td style="text-align: left;">Luk_TRB_EBV_CD8_3</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">39014</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
</tbody>
</table>
<ul>
<li><code>count_table</code> which can be created with <a href="../usage_intersections/#count_table">Intersections</a> <code>count_table</code> (for clonotypes of <a href="./#for-clonotype-groups">clonotype groups</a>) or <code>count_table_by_cluster</code> (for <a href="./#count-table-by-cluster">clusters</a>) functions. Example:</li>
</ul>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">feature_id</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_3</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">cluster_0</td>
<td style="text-align: right;">0.329744</td>
<td style="text-align: right;">0.323807</td>
<td style="text-align: right;">0.31638</td>
<td style="text-align: right;">0.178582</td>
<td style="text-align: right;">0.190506</td>
<td style="text-align: right;">0.183408</td>
<td style="text-align: right;">0.194283</td>
<td style="text-align: right;">0.314896</td>
<td style="text-align: right;">0.266949</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">cluster_1</td>
<td style="text-align: right;">0.0875153</td>
<td style="text-align: right;">0.0826597</td>
<td style="text-align: right;">0.12148</td>
<td style="text-align: right;">0.0544798</td>
<td style="text-align: right;">0.0355812</td>
<td style="text-align: right;">0.0637625</td>
<td style="text-align: right;">0.0852074</td>
<td style="text-align: right;">0.0706873</td>
<td style="text-align: right;">0.075124</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">cluster_2</td>
<td style="text-align: right;">0.0126884</td>
<td style="text-align: right;">0.0273287</td>
<td style="text-align: right;">0.0147922</td>
<td style="text-align: right;">0.386493</td>
<td style="text-align: right;">0.449984</td>
<td style="text-align: right;">0.361615</td>
<td style="text-align: right;">0.0247404</td>
<td style="text-align: right;">0.0136364</td>
<td style="text-align: right;">0.0211886</td>
</tr>
<tr>
<td style="text-align: right;">3</td>
<td style="text-align: left;">cluster_3</td>
<td style="text-align: right;">0.20035</td>
<td style="text-align: right;">0.172857</td>
<td style="text-align: right;">0.192064</td>
<td style="text-align: right;">0.0404682</td>
<td style="text-align: right;">0.0345952</td>
<td style="text-align: right;">0.0217206</td>
<td style="text-align: right;">0.214885</td>
<td style="text-align: right;">0.180312</td>
<td style="text-align: right;">0.163982</td>
</tr>
<tr>
<td style="text-align: right;">4</td>
<td style="text-align: left;">cluster_4</td>
<td style="text-align: right;">0.00825616</td>
<td style="text-align: right;">0.0143059</td>
<td style="text-align: right;">0.0245092</td>
<td style="text-align: right;">0.00313302</td>
<td style="text-align: right;">0.0107374</td>
<td style="text-align: right;">0.00292042</td>
<td style="text-align: right;">0.00648297</td>
<td style="text-align: right;">0.00469606</td>
<td style="text-align: right;">0.00751628</td>
</tr>
</tbody>
</table>
<div class="admonition tip">
<p class="admonition-title">Tips</p>
<p>For the <code>count_table</code>, itâ€™s best to use UMI counts. If deep sequencing samples (&gt;30 000 UMIs per sample) are available, it is recommended to downsample them to the same number of UMIs for normalization and faster computation. For details on downsampling filter usage, visit <a href="../usage_stats/#filtering-clonosets">stats usage</a> page.</p>
</div>
<p>Also, <code>clonosets_df</code> is required for creating a count table. Example (here, a subset of CD8 cells is selected):</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="n">count_table_by_cluster_cd8</span>
</code></pre></div>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">chain</th>
<th style="text-align: left;">filename</th>
<th style="text-align: left;">Subset</th>
<th style="text-align: right;">Replica</th>
<th style="text-align: right;">Number_of_cells</th>
<th style="text-align: left;">Peptid</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">Luk_TRB_AdV_CD8_3</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_0</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">21031</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">Luk_TRB_CMV_CD8_2</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_1</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">17490</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">3</td>
<td style="text-align: left;">Luk_TRB_AdV_CD8_1</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_2</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">37698</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">5</td>
<td style="text-align: left;">Luk_TRB_CMV_CD8_3</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_3</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">16662</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
<tr>
<td style="text-align: right;">6</td>
<td style="text-align: left;">Luk_TRB_EBV_CD8_2</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_4</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">22020</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">12</td>
<td style="text-align: left;">Luk_TRB_EBV_CD8_1</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_5</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">41274</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">14</td>
<td style="text-align: left;">Luk_TRB_EBV_CD8_3</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_6</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">39014</td>
<td style="text-align: left;">PepTivator_EBV</td>
</tr>
<tr>
<td style="text-align: right;">16</td>
<td style="text-align: left;">Luk_TRB_AdV_CD8_2</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_7</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">21359</td>
<td style="text-align: left;">PepTivator_AdV</td>
</tr>
<tr>
<td style="text-align: right;">17</td>
<td style="text-align: left;">Luk_TRB_CMV_CD8_1</td>
<td style="text-align: left;">TRB</td>
<td style="text-align: left;">path_to_sample_8</td>
<td style="text-align: left;">CD8</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">19617</td>
<td style="text-align: left;">PepTivator_CMV</td>
</tr>
</tbody>
</table>
<p><br></p>
<h2 id="count-table-by-cluster">Count table by cluster</h2>
<h3 id="creating-clusters">Creating clusters</h3>
<p>Here, only functional clonotypes are used (no frameshifts or premature stop codons) and one-node clusters are filtered out. </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="n">func_filter</span> <span class="o">=</span> <span class="n">clf</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="n">functionality</span><span class="o">=</span><span class="s2">&quot;f&quot;</span><span class="p">,</span> <span class="n">by_umi</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="n">clusters_cd8</span> <span class="o">=</span> <span class="n">clustering</span><span class="o">.</span><span class="n">create_clusters</span><span class="p">(</span><span class="n">cd8_clonosets</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">func_filter</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">igh</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">tcrdist_radius</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">count_by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="n">clusters_cd8_nonsingle</span> <span class="o">=</span> <span class="n">clustering</span><span class="o">.</span><span class="n">filter_one_node_clusters</span><span class="p">(</span><span class="n">clusters_cd8</span><span class="p">)</span>
</code></pre></div>
<h3 id="count_table_by_cluster"><code>count_table_by_cluster</code></h3>
<div class="admonition tip">
<p class="admonition-title">Number of mismatches</p>
<p>It is recommended to set <code>mismatches</code> to 0, as the results with either 0 or 1 <code>mismatches</code> will be identical, however, using the former will result in faster computation. When the same clonoset is used for both clustering and intersections, the clone will be already present in the cluster.</p>
</div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="n">count_table_by_cluster_cd8</span> <span class="o">=</span> <span class="n">intersections</span><span class="o">.</span><span class="n">count_table_by_cluster</span><span class="p">(</span><span class="n">cd8_clonosets</span><span class="p">,</span> <span class="n">clusters_cd8_nonsingle</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">func_filter</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="n">count_table_by_cluster_cd8</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">feature_id</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_3</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">cluster_0</td>
<td style="text-align: right;">0.329744</td>
<td style="text-align: right;">0.323807</td>
<td style="text-align: right;">0.31638</td>
<td style="text-align: right;">0.178582</td>
<td style="text-align: right;">0.190506</td>
<td style="text-align: right;">0.183408</td>
<td style="text-align: right;">0.194283</td>
<td style="text-align: right;">0.314896</td>
<td style="text-align: right;">0.266949</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">cluster_1</td>
<td style="text-align: right;">0.0875153</td>
<td style="text-align: right;">0.0826597</td>
<td style="text-align: right;">0.12148</td>
<td style="text-align: right;">0.0544798</td>
<td style="text-align: right;">0.0355812</td>
<td style="text-align: right;">0.0637625</td>
<td style="text-align: right;">0.0852074</td>
<td style="text-align: right;">0.0706873</td>
<td style="text-align: right;">0.075124</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">cluster_2</td>
<td style="text-align: right;">0.0126884</td>
<td style="text-align: right;">0.0273287</td>
<td style="text-align: right;">0.0147922</td>
<td style="text-align: right;">0.386493</td>
<td style="text-align: right;">0.449984</td>
<td style="text-align: right;">0.361615</td>
<td style="text-align: right;">0.0247404</td>
<td style="text-align: right;">0.0136364</td>
<td style="text-align: right;">0.0211886</td>
</tr>
<tr>
<td style="text-align: right;">3</td>
<td style="text-align: left;">cluster_3</td>
<td style="text-align: right;">0.20035</td>
<td style="text-align: right;">0.172857</td>
<td style="text-align: right;">0.192064</td>
<td style="text-align: right;">0.0404682</td>
<td style="text-align: right;">0.0345952</td>
<td style="text-align: right;">0.0217206</td>
<td style="text-align: right;">0.214885</td>
<td style="text-align: right;">0.180312</td>
<td style="text-align: right;">0.163982</td>
</tr>
<tr>
<td style="text-align: right;">4</td>
<td style="text-align: left;">cluster_4</td>
<td style="text-align: right;">0.00825616</td>
<td style="text-align: right;">0.0143059</td>
<td style="text-align: right;">0.0245092</td>
<td style="text-align: right;">0.00313302</td>
<td style="text-align: right;">0.0107374</td>
<td style="text-align: right;">0.00292042</td>
<td style="text-align: right;">0.00648297</td>
<td style="text-align: right;">0.00469606</td>
<td style="text-align: right;">0.00751628</td>
</tr>
</tbody>
</table>
<p><br></p>
<h2 id="differential-expression-calculation">Differential expression calculation</h2>
<p><code>wilcox_diff_expression</code> has two operating modes: </p>
<ul>
<li>If there are 2 group present, a pairwise comparison is done;</li>
<li>If the number of groups is higher than 2, each group is compared to all others, with additional columns for pairwise comparison between groups.</li>
</ul>
<p>Features (clonotypes or clusters) that appear in <code>min_samples</code> or more samples and have a <code>count_threshold</code> or higher counts or frequencies. This filters out unreliable differences, as the final results undergo FDR correction by the BH method. The default values for <code>min_samples</code> and <code>count_threshold</code> are <code>2</code> and <code>5e-04</code>, respectively. </p>
<h3 id="for-clusters">For clusters</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="n">diff_expression_by_cluster</span> <span class="o">=</span> <span class="n">diffexp</span><span class="o">.</span><span class="n">wilcox_diff_expression</span><span class="p">(</span><span class="n">count_table_by_cluster_cd8</span><span class="p">,</span> <span class="n">sample_metadata</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">count_threshold</span><span class="o">=</span><span class="mf">0.00005</span><span class="p">)</span>
</code></pre></div>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">feature_id</th>
<th style="text-align: right;">logFC</th>
<th style="text-align: right;">U</th>
<th style="text-align: right;">p</th>
<th style="text-align: right;">p_adj</th>
<th style="text-align: left;">pair</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_3</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">959</td>
<td style="text-align: left;">cluster_71</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.000304934</td>
<td style="text-align: right;">0.0227611</td>
<td style="text-align: left;">PepTivator_EBV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.00337752</td>
<td style="text-align: right;">0.00498357</td>
<td style="text-align: right;">0.000248729</td>
</tr>
<tr>
<td style="text-align: right;">1007</td>
<td style="text-align: left;">cluster_76</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.000304934</td>
<td style="text-align: right;">0.0227611</td>
<td style="text-align: left;">PepTivator_EBV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.000309606</td>
<td style="text-align: right;">0.000533954</td>
<td style="text-align: right;">0.000287593</td>
</tr>
<tr>
<td style="text-align: right;">1072</td>
<td style="text-align: left;">cluster_83</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.000304934</td>
<td style="text-align: right;">0.0227611</td>
<td style="text-align: left;">PepTivator_CMV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0018276</td>
<td style="text-align: right;">0.00372521</td>
<td style="text-align: right;">0.00638842</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<h3 id="for-clonotype-groups">For clonotype groups</h3>
<p>Here, a clonotype group is defined based on the <code>overlap_type</code> used during <code>count_table</code> calculation. <br>Possible overlap types are [<code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>], <code>aa</code>/<code>nt</code> stands for an amino acid or nucleotide sequence, and <code>V</code>/<code>J</code>/<code>VJ</code> denote a segment type. </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="n">count_table_by_clonotypes_aaV_cd8</span> <span class="o">=</span> <span class="n">intersections</span><span class="o">.</span><span class="n">count_table</span><span class="p">(</span><span class="n">cd8_clonosets</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">func_filter</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
</code></pre></div>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">feature_id</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_3</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">('CASSPGLAAPYSEQFF', 'TRBV12-3')</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">('CASRQGAGEQYF', 'TRBV19')</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">('CSAGTTYGTDIISQHF', 'TRBV20-1')</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="n">diff_expression_by_clonotypes_aaV</span> <span class="o">=</span> <span class="n">diffexp</span><span class="o">.</span><span class="n">wilcox_diff_expression</span><span class="p">(</span><span class="n">count_table_by_clonotypes_aaV_cd8</span><span class="p">,</span> <span class="n">sample_metadata</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">count_threshold</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</code></pre></div>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">feature_id</th>
<th style="text-align: right;">logFC</th>
<th style="text-align: right;">U</th>
<th style="text-align: right;">p</th>
<th style="text-align: right;">p_adj</th>
<th style="text-align: left;">pair</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_3</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">('CASSSGLLGEQFF', 'TRBV6-5')</td>
<td style="text-align: right;">9.62006</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.00118822</td>
<td style="text-align: right;">0.0481036</td>
<td style="text-align: left;">PepTivator_EBV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">421</td>
<td style="text-align: right;">204</td>
<td style="text-align: right;">3141</td>
</tr>
<tr>
<td style="text-align: right;">111</td>
<td style="text-align: left;">('CASSSGLLDTQYF', 'TRBV6-5')</td>
<td style="text-align: right;">7.93852</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.00118822</td>
<td style="text-align: right;">0.0481036</td>
<td style="text-align: left;">PepTivator_EBV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1428</td>
<td style="text-align: right;">1462</td>
<td style="text-align: right;">614</td>
</tr>
<tr>
<td style="text-align: right;">118</td>
<td style="text-align: left;">('CASSRTSGSFLFEQYF', 'TRBV14')</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.000304934</td>
<td style="text-align: right;">0.0201417</td>
<td style="text-align: left;">PepTivator_AdV_vs_all</td>
<td style="text-align: right;">223</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<p>Output has the following columns:</p>
<ul>
<li>
<p><b>LogFC</b>: Log fold change of counts / frequencies between two groups. If all the samples in one group have zero counts or frequencies, logFC is set to 100 or -100.</p>
</li>
<li>
<p><b>U</b>: Mann-Whitney U statistic</p>
</li>
<li>
<p><b>p</b>: p-value</p>
</li>
<li>
<p><b>p_adj</b>: p-value after multiple testing correction using Benjamini-Hochberg (BH) procedure</p>
</li>
<li>
<p>Pairwise group comparisons and their raw p-values (_p in a column name) and log fold changes (_logFC in a column name).</p>
</li>
</ul>
<p><br></p>
<h2 id="merge-cluster-properties-by-feature_id">Merge cluster properties by <code>feature_id</code></h2>
<ul>
<li>Create cluster properties</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="n">cd8_cluster_props</span> <span class="o">=</span> <span class="n">clustering</span><span class="o">.</span><span class="n">cluster_properties</span><span class="p">(</span><span class="n">clusters_cd8_nonsingle</span><span class="p">,</span> <span class="n">weighed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cluster_id&quot;</span><span class="p">:</span> <span class="s2">&quot;feature_id&quot;</span><span class="p">})</span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="n">diff_expression_by_cluster</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cd8_cluster_props</span><span class="p">)</span>
</code></pre></div>
<table>
<thead>
<tr>
<th style="text-align: right;"></th>
<th style="text-align: left;">feature_id</th>
<th style="text-align: right;">logFC</th>
<th style="text-align: right;">U</th>
<th style="text-align: right;">p</th>
<th style="text-align: right;">p_adj</th>
<th style="text-align: left;">pair</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_AdV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_CMV_CD8_3</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_1</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_2</th>
<th style="text-align: right;">Luk_TRB_EBV_CD8_3</th>
<th style="text-align: right;">nodes</th>
<th style="text-align: right;">edges</th>
<th style="text-align: right;">diameter</th>
<th style="text-align: right;">density</th>
<th style="text-align: right;">eccentricity</th>
<th style="text-align: left;">concensus_cdr3aa</th>
<th style="text-align: left;">concensus_cdr3nt</th>
<th style="text-align: left;">concensus_v</th>
<th style="text-align: left;">concensus_j</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">cluster_71</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.000304934</td>
<td style="text-align: right;">0.0227611</td>
<td style="text-align: left;">PepTivator_EBV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.00337752</td>
<td style="text-align: right;">0.00498357</td>
<td style="text-align: right;">0.000248729</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">CASSGASGSFNEQFF</td>
<td style="text-align: left;">TGTGCCAGTAGTGGGGCTAGCGGGAGTTTTAATGAGCAGTTCTTC</td>
<td style="text-align: left;">TRBV19</td>
<td style="text-align: left;">TRBJ2-1</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">cluster_76</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.000304934</td>
<td style="text-align: right;">0.0227611</td>
<td style="text-align: left;">PepTivator_EBV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.000309606</td>
<td style="text-align: right;">0.000533954</td>
<td style="text-align: right;">0.000287593</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">21</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">CASTSHGTSKDNEQFF</td>
<td style="text-align: left;">TGTGCCAGCACGTCTCACGGGACTAGCAAAGACAATGAGCAGTTCTTC</td>
<td style="text-align: left;">TRBV7-6</td>
<td style="text-align: left;">TRBJ2-1</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">cluster_83</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">0.000304934</td>
<td style="text-align: right;">0.0227611</td>
<td style="text-align: left;">PepTivator_CMV_vs_all</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0018276</td>
<td style="text-align: right;">0.00372521</td>
<td style="text-align: right;">0.00638842</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">14</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0.933333</td>
<td style="text-align: right;">1.33333</td>
<td style="text-align: left;">CASSFLLGQGADYEQYF</td>
<td style="text-align: left;">TGTGCCAGCAGCTTCCTCCTGGGACAGGGGGCCGACTACGAGCAGTACTTC</td>
<td style="text-align: left;">TRBV11-2</td>
<td style="text-align: left;">TRBJ2-7</td>
</tr>
</tbody>
</table>
<details class="info">
<summary>Visualization</summary>
</details>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2023 <a href="https://github.com/mmjmike" target="_blank" rel="noopener">Mikhail Myshkin</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/mmjmike" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M202.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M496 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.path", "toc.integrate", "navigation.top", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>