
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>Page 2 - Repseq Library for Immune repertoire postanalysis</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#page-2" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-header__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Repseq Library for Immune repertoire postanalysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Page 2
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../installation/" class="md-tabs__link">
        
  
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../modules/" class="md-tabs__link">
        
  
  
    
  
  Modules

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../usage_mixcr/" class="md-tabs__link">
          
  
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../functions/" class="md-tabs__link">
        
  
  
    
  
  Functions

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Repseq Library for Immune repertoire postanalysis" class="md-nav__button md-logo" aria-label="Repseq Library for Immune repertoire postanalysis" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Repseq Library for Immune repertoire postanalysis
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Modules
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Usage
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_mixcr/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MiXCR
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_stats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stats
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_intersections/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Intersections
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_clustering/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Clustering
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage_diffexp/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Differential expression
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Functions
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              
                



  


              
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="page-2">Page 2</h1>
<h2 id="code-annotation-examples">Code annotation examples</h2>
<h3 id="code-blocks">Code blocks</h3>
<p>Some <code>code</code> goes here</p>
<h3 id="plain-code-blocks">Plain code blocks</h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>def read_json_report(sample_id, folder, report_type):
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>    filename = os.path.join(folder, f&quot;{sample_id}.{report_type}.report.json&quot;)
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>    with open(filename) as data_file:    
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>        for jsonObj in data_file:
<a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>            report = json.loads(jsonObj)
<a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>    return report
<a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a>// some comment
</code></pre></div>
<h4 id="code-for-a-specific-language">Code for a specific language</h4>
<p>Some more code with the <code>py</code> at the start</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="n">REPSEQ_PATH</span> <span class="o">=</span> <span class="s1">&#39;/home/mmyshkin/soft/repseq&#39;</span>
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">REPSEQ_PATH</span><span class="p">)</span>
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">slurm</span>
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">io</span>
<a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">common_functions</span> <span class="k">as</span> <span class="n">cf</span>
<a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">clonosets</span> <span class="k">as</span> <span class="n">cl</span>
<a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">clustering</span>
<a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">mixcr</span> <span class="k">as</span> <span class="n">mx</span>
<a id="__codelineno-1-10" name="__codelineno-1-10" href="#__codelineno-1-10"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">segment_usage</span> <span class="k">as</span> <span class="n">su</span>
<a id="__codelineno-1-11" name="__codelineno-1-11" href="#__codelineno-1-11"></a><span class="kn">from</span><span class="w"> </span><span class="nn">repseq</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
</code></pre></div>
<h4 id="code-with-a-title">Code with a title</h4>
<div class="highlight"><span class="filename">import useful packages</span><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">Image</span><span class="p">,</span> <span class="n">display</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a><span class="kn">import</span><span class="w"> </span><span class="nn">math</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a><span class="kn">import</span><span class="w"> </span><span class="nn">random</span>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
</code></pre></div>
<h4 id="add-line-numbers">Add line numbers</h4>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-3-1">1</a></span>
<span class="normal"><a href="#__codelineno-3-2">2</a></span>
<span class="normal"><a href="#__codelineno-3-3">3</a></span>
<span class="normal"><a href="#__codelineno-3-4">4</a></span>
<span class="normal"><a href="#__codelineno-3-5">5</a></span>
<span class="normal"><a href="#__codelineno-3-6">6</a></span>
<span class="normal"><a href="#__codelineno-3-7">7</a></span>
<span class="normal"><a href="#__codelineno-3-8">8</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1"></a><span class="k">def</span><span class="w"> </span><span class="nf">shannon_wiener</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">):</span>
<a id="__codelineno-3-2" name="__codelineno-3-2"></a>    <span class="n">list_of_numbers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-3-3" name="__codelineno-3-3"></a>    <span class="n">total_size</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-3-4" name="__codelineno-3-4"></a>    <span class="n">freqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">/</span><span class="n">total_size</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">list_of_numbers</span><span class="p">]</span>
<a id="__codelineno-3-5" name="__codelineno-3-5"></a>    <span class="n">diversity</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-3-6" name="__codelineno-3-6"></a>    <span class="n">sw</span> <span class="o">=</span> <span class="o">-</span><span class="nb">sum</span><span class="p">([</span><span class="n">f</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">])</span>
<a id="__codelineno-3-7" name="__codelineno-3-7"></a>    <span class="n">sw_norm</span> <span class="o">=</span> <span class="n">sw</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">diversity</span><span class="p">)</span>
<a id="__codelineno-3-8" name="__codelineno-3-8"></a>    <span class="k">return</span> <span class="n">sw</span><span class="p">,</span> <span class="n">sw_norm</span><span class="p">,</span> <span class="n">diversity</span>
</code></pre></div></td></tr></table></div>
<h4 id="highlighting-lines">Highlighting lines</h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="k">def</span><span class="w"> </span><span class="nf">shannon_wiener</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">):</span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>    <span class="n">list_of_numbers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>    <span class="n">total_size</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>    <span class="n">freqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">/</span><span class="n">total_size</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">list_of_numbers</span><span class="p">]</span>
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a><span class="hll">    <span class="n">diversity</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_of_numbers</span><span class="p">)</span>
</span><a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a><span class="hll">    <span class="n">sw</span> <span class="o">=</span> <span class="o">-</span><span class="nb">sum</span><span class="p">([</span><span class="n">f</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">freqs</span><span class="p">])</span>
</span><a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a><span class="hll">    <span class="n">sw_norm</span> <span class="o">=</span> <span class="n">sw</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">diversity</span><span class="p">)</span>
</span><a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a>    <span class="k">return</span> <span class="n">sw</span><span class="p">,</span> <span class="n">sw_norm</span><span class="p">,</span> <span class="n">diversity</span>
</code></pre></div>
<h2 id="icons-and-emojis">Icons and Emojis</h2>
<p><img alt="ðŸ˜„" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/1f604.svg" title=":smile:" /></p>
<p><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M464 256a208 208 0 1 0-416 0 208 208 0 1 0 416 0M0 256a256 256 0 1 1 512 0 256 256 0 1 1-512 0m118.3 58.2c-4.2-13.7 7.1-26.2 21.4-26.2h232.6c14.3 0 25.6 12.5 21.4 26.2C375.7 373.1 320.8 416 256 416s-119.7-42.9-137.7-101.8M144 192a32 32 0 1 1 64 0 32 32 0 1 1-64 0m164 8c0 11-9 20-20 20s-20-9-20-20c0-33.1 26.9-60 60-60h16c33.1 0 60 26.9 60 60 0 11-9 20-20 20s-20-9-20-20-9-20-20-20h-16c-11 0-20 9-20 20"/></svg></span></p>
<p><span class="twemoji twitter"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M459.4 151.7c.3 4.5.3 9.1.3 13.6 0 138.7-105.6 298.6-298.6 298.6-59.5 0-114.7-17.2-161.1-47.1 8.4 1 16.6 1.3 25.3 1.3 49.1 0 94.2-16.6 130.3-44.8-46.1-1-84.8-31.2-98.1-72.8 6.5 1 13 1.6 19.8 1.6 9.4 0 18.8-1.3 27.6-3.6-48.1-9.7-84.1-52-84.1-103v-1.3c14 7.8 30.2 12.7 47.4 13.3-28.3-18.8-46.8-51-46.8-87.4 0-19.5 5.2-37.4 14.3-53C87.4 130.8 165 172.4 252.1 176.9c-1.6-7.8-2.6-15.9-2.6-24C249.5 95.1 296.3 48 354.4 48c30.2 0 57.5 12.7 76.7 33.1 23.7-4.5 46.5-13.3 66.6-25.3-7.8 24.4-24.4 44.8-46.1 57.8 21.1-2.3 41.6-8.1 60.4-16.2-14.3 20.8-32.2 39.3-52.6 54.3"/></svg></span></p>
<p><span class="twemoji heart"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14 20.408c-.492.308-.903.546-1.192.709q-.23.129-.463.252h-.002a.75.75 0 0 1-.686 0 17 17 0 0 1-.465-.252 31 31 0 0 1-4.803-3.34C3.8 15.572 1 12.331 1 8.513 1 5.052 3.829 2.5 6.736 2.5 9.03 2.5 10.881 3.726 12 5.605 13.12 3.726 14.97 2.5 17.264 2.5 20.17 2.5 23 5.052 23 8.514c0 3.818-2.801 7.06-5.389 9.262A31 31 0 0 1 14 20.408"/></svg></span></p>
<h2 id="try-to-show-all-the-functions-of-a-module">Try to show all the functions of a module</h2>


<div class="doc doc-object doc-module">



<a id="intersections"></a>
    <div class="doc doc-contents first">










<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="intersections.count_table" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">count_table</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">strict_presence</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Creates a table that shows how many times each unique clonotype appears across different clonosets. It processes a given dataset of clonotypes (clonosets_df) 
and generates a frequency/count table based on a specified overlap type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
<code>filename</code> - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>overlap_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mismatches</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Max number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>only_functional</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>strict_presence (bool, default</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>): if set to <code>True</code> and the clonotype is not found in the clonoset, it will not be counted, even when the <code>mismatches</code> option is not set to 0.
If <code>False</code>, mismatched sequences are counted even if the exact match does not exist.
    by_freq (bool): default is <code>True</code> - this means that the intersect metric is frequency of clonotype, 
but not its count</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>by_freq</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>default is <code>True</code> - this means that the intersect metric is frequency of clonotype, 
but not its count</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe containing clonotype counts with sample names as columns and all possible clonotypes (given the overlap_type) as rows.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-83"> 83</a></span>
<span class="normal"><a href="#__codelineno-0-84"> 84</a></span>
<span class="normal"><a href="#__codelineno-0-85"> 85</a></span>
<span class="normal"><a href="#__codelineno-0-86"> 86</a></span>
<span class="normal"><a href="#__codelineno-0-87"> 87</a></span>
<span class="normal"><a href="#__codelineno-0-88"> 88</a></span>
<span class="normal"><a href="#__codelineno-0-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-0-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-0-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-0-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-0-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-0-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-0-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-0-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-0-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-0-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-0-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-0-100">100</a></span>
<span class="normal"><a href="#__codelineno-0-101">101</a></span>
<span class="normal"><a href="#__codelineno-0-102">102</a></span>
<span class="normal"><a href="#__codelineno-0-103">103</a></span>
<span class="normal"><a href="#__codelineno-0-104">104</a></span>
<span class="normal"><a href="#__codelineno-0-105">105</a></span>
<span class="normal"><a href="#__codelineno-0-106">106</a></span>
<span class="normal"><a href="#__codelineno-0-107">107</a></span>
<span class="normal"><a href="#__codelineno-0-108">108</a></span>
<span class="normal"><a href="#__codelineno-0-109">109</a></span>
<span class="normal"><a href="#__codelineno-0-110">110</a></span>
<span class="normal"><a href="#__codelineno-0-111">111</a></span>
<span class="normal"><a href="#__codelineno-0-112">112</a></span>
<span class="normal"><a href="#__codelineno-0-113">113</a></span>
<span class="normal"><a href="#__codelineno-0-114">114</a></span>
<span class="normal"><a href="#__codelineno-0-115">115</a></span>
<span class="normal"><a href="#__codelineno-0-116">116</a></span>
<span class="normal"><a href="#__codelineno-0-117">117</a></span>
<span class="normal"><a href="#__codelineno-0-118">118</a></span>
<span class="normal"><a href="#__codelineno-0-119">119</a></span>
<span class="normal"><a href="#__codelineno-0-120">120</a></span>
<span class="normal"><a href="#__codelineno-0-121">121</a></span>
<span class="normal"><a href="#__codelineno-0-122">122</a></span>
<span class="normal"><a href="#__codelineno-0-123">123</a></span>
<span class="normal"><a href="#__codelineno-0-124">124</a></span>
<span class="normal"><a href="#__codelineno-0-125">125</a></span>
<span class="normal"><a href="#__codelineno-0-126">126</a></span>
<span class="normal"><a href="#__codelineno-0-127">127</a></span>
<span class="normal"><a href="#__codelineno-0-128">128</a></span>
<span class="normal"><a href="#__codelineno-0-129">129</a></span>
<span class="normal"><a href="#__codelineno-0-130">130</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-83" name="__codelineno-0-83"></a><span class="k">def</span><span class="w"> </span><span class="nf">count_table</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">strict_presence</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-84" name="__codelineno-0-84"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-85" name="__codelineno-0-85"></a><span class="sd">    Creates a table that shows how many times each unique clonotype appears across different clonosets. It processes a given dataset of clonotypes (clonosets_df) </span>
<a id="__codelineno-0-86" name="__codelineno-0-86"></a><span class="sd">    and generates a frequency/count table based on a specified overlap type.</span>
<a id="__codelineno-0-87" name="__codelineno-0-87"></a>
<a id="__codelineno-0-88" name="__codelineno-0-88"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-89" name="__codelineno-0-89"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-90" name="__codelineno-0-90"></a><span class="sd">            `filename` - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-91" name="__codelineno-0-91"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-92" name="__codelineno-0-92"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-93" name="__codelineno-0-93"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-94" name="__codelineno-0-94"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-95" name="__codelineno-0-95"></a><span class="sd">        mismatches (int): Max number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-96" name="__codelineno-0-96"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-97" name="__codelineno-0-97"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-98" name="__codelineno-0-98"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-99" name="__codelineno-0-99"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-100" name="__codelineno-0-100"></a><span class="sd">        strict_presence (bool, default:): if set to `True` and the clonotype is not found in the clonoset, it will not be counted, even when the `mismatches` option is not set to 0.</span>
<a id="__codelineno-0-101" name="__codelineno-0-101"></a><span class="sd">            If `False`, mismatched sequences are counted even if the exact match does not exist.</span>
<a id="__codelineno-0-102" name="__codelineno-0-102"></a><span class="sd">                by_freq (bool): default is `True` - this means that the intersect metric is frequency of clonotype, </span>
<a id="__codelineno-0-103" name="__codelineno-0-103"></a><span class="sd">            but not its count</span>
<a id="__codelineno-0-104" name="__codelineno-0-104"></a><span class="sd">        by_freq (bool): default is `True` - this means that the intersect metric is frequency of clonotype, </span>
<a id="__codelineno-0-105" name="__codelineno-0-105"></a><span class="sd">            but not its count</span>
<a id="__codelineno-0-106" name="__codelineno-0-106"></a>
<a id="__codelineno-0-107" name="__codelineno-0-107"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-108" name="__codelineno-0-108"></a><span class="sd">        df (pd.DataFrame): dataframe containing clonotype counts with sample names as columns and all possible clonotypes (given the overlap_type) as rows. </span>
<a id="__codelineno-0-109" name="__codelineno-0-109"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-110" name="__codelineno-0-110"></a>
<a id="__codelineno-0-111" name="__codelineno-0-111"></a>
<a id="__codelineno-0-112" name="__codelineno-0-112"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating clonotypes count table</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-113" name="__codelineno-0-113"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-114" name="__codelineno-0-114"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-115" name="__codelineno-0-115"></a>    <span class="n">clonoset_dicts</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">,</span>
<a id="__codelineno-0-116" name="__codelineno-0-116"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-117" name="__codelineno-0-117"></a>    <span class="n">unique_clonotypes</span> <span class="o">=</span> <span class="n">find_unique_clonotypes_in_clonoset_dicts</span><span class="p">(</span><span class="n">clonoset_dicts</span><span class="p">)</span>
<a id="__codelineno-0-118" name="__codelineno-0-118"></a>
<a id="__codelineno-0-119" name="__codelineno-0-119"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-120" name="__codelineno-0-120"></a>    <span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">clonoset_dicts</span><span class="p">:</span>
<a id="__codelineno-0-121" name="__codelineno-0-121"></a>        <span class="n">task</span> <span class="o">=</span> <span class="p">[</span><span class="n">unique_clonotypes</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">clonoset_dicts</span><span class="p">[</span><span class="n">sample_id</span><span class="p">],</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">strict_presence</span><span class="p">]</span>
<a id="__codelineno-0-122" name="__codelineno-0-122"></a>        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
<a id="__codelineno-0-123" name="__codelineno-0-123"></a>
<a id="__codelineno-0-124" name="__codelineno-0-124"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">count_table_mp</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Counting features&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;clonosets&quot;</span><span class="p">)</span>
<a id="__codelineno-0-125" name="__codelineno-0-125"></a>    <span class="n">result_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-126" name="__codelineno-0-126"></a>    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
<a id="__codelineno-0-127" name="__codelineno-0-127"></a>        <span class="n">result_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<a id="__codelineno-0-128" name="__codelineno-0-128"></a>    <span class="n">count_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_dict</span><span class="p">)</span>
<a id="__codelineno-0-129" name="__codelineno-0-129"></a>    <span class="n">count_table</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">unique_clonotypes</span>
<a id="__codelineno-0-130" name="__codelineno-0-130"></a>    <span class="k">return</span> <span class="n">count_table</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.count_table_by_cluster" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">count_table_by_cluster</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clusters_list</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>This function creates a table that shows the presence of clonotypes grouped into user-provided clusters across different clonosets. Instead of counting individual clonotypes, it 
calculates how many clonotypes from each cluster appear in each clonoset.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cluster_list</code>
            </td>
            <td>
                  <code>?</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>description</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>overlap_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mismatches</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Max number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>only_functional</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with the following columns: description</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-162">162</a></span>
<span class="normal"><a href="#__codelineno-0-163">163</a></span>
<span class="normal"><a href="#__codelineno-0-164">164</a></span>
<span class="normal"><a href="#__codelineno-0-165">165</a></span>
<span class="normal"><a href="#__codelineno-0-166">166</a></span>
<span class="normal"><a href="#__codelineno-0-167">167</a></span>
<span class="normal"><a href="#__codelineno-0-168">168</a></span>
<span class="normal"><a href="#__codelineno-0-169">169</a></span>
<span class="normal"><a href="#__codelineno-0-170">170</a></span>
<span class="normal"><a href="#__codelineno-0-171">171</a></span>
<span class="normal"><a href="#__codelineno-0-172">172</a></span>
<span class="normal"><a href="#__codelineno-0-173">173</a></span>
<span class="normal"><a href="#__codelineno-0-174">174</a></span>
<span class="normal"><a href="#__codelineno-0-175">175</a></span>
<span class="normal"><a href="#__codelineno-0-176">176</a></span>
<span class="normal"><a href="#__codelineno-0-177">177</a></span>
<span class="normal"><a href="#__codelineno-0-178">178</a></span>
<span class="normal"><a href="#__codelineno-0-179">179</a></span>
<span class="normal"><a href="#__codelineno-0-180">180</a></span>
<span class="normal"><a href="#__codelineno-0-181">181</a></span>
<span class="normal"><a href="#__codelineno-0-182">182</a></span>
<span class="normal"><a href="#__codelineno-0-183">183</a></span>
<span class="normal"><a href="#__codelineno-0-184">184</a></span>
<span class="normal"><a href="#__codelineno-0-185">185</a></span>
<span class="normal"><a href="#__codelineno-0-186">186</a></span>
<span class="normal"><a href="#__codelineno-0-187">187</a></span>
<span class="normal"><a href="#__codelineno-0-188">188</a></span>
<span class="normal"><a href="#__codelineno-0-189">189</a></span>
<span class="normal"><a href="#__codelineno-0-190">190</a></span>
<span class="normal"><a href="#__codelineno-0-191">191</a></span>
<span class="normal"><a href="#__codelineno-0-192">192</a></span>
<span class="normal"><a href="#__codelineno-0-193">193</a></span>
<span class="normal"><a href="#__codelineno-0-194">194</a></span>
<span class="normal"><a href="#__codelineno-0-195">195</a></span>
<span class="normal"><a href="#__codelineno-0-196">196</a></span>
<span class="normal"><a href="#__codelineno-0-197">197</a></span>
<span class="normal"><a href="#__codelineno-0-198">198</a></span>
<span class="normal"><a href="#__codelineno-0-199">199</a></span>
<span class="normal"><a href="#__codelineno-0-200">200</a></span>
<span class="normal"><a href="#__codelineno-0-201">201</a></span>
<span class="normal"><a href="#__codelineno-0-202">202</a></span>
<span class="normal"><a href="#__codelineno-0-203">203</a></span>
<span class="normal"><a href="#__codelineno-0-204">204</a></span>
<span class="normal"><a href="#__codelineno-0-205">205</a></span>
<span class="normal"><a href="#__codelineno-0-206">206</a></span>
<span class="normal"><a href="#__codelineno-0-207">207</a></span>
<span class="normal"><a href="#__codelineno-0-208">208</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-162" name="__codelineno-0-162"></a><span class="k">def</span><span class="w"> </span><span class="nf">count_table_by_cluster</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clusters_list</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<a id="__codelineno-0-163" name="__codelineno-0-163"></a>
<a id="__codelineno-0-164" name="__codelineno-0-164"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-165" name="__codelineno-0-165"></a><span class="sd">    This function creates a table that shows the presence of clonotypes grouped into user-provided clusters across different clonosets. Instead of counting individual clonotypes, it </span>
<a id="__codelineno-0-166" name="__codelineno-0-166"></a><span class="sd">    calculates how many clonotypes from each cluster appear in each clonoset.</span>
<a id="__codelineno-0-167" name="__codelineno-0-167"></a>
<a id="__codelineno-0-168" name="__codelineno-0-168"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-169" name="__codelineno-0-169"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-170" name="__codelineno-0-170"></a><span class="sd">            filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-171" name="__codelineno-0-171"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-172" name="__codelineno-0-172"></a><span class="sd">        cluster_list (?): description</span>
<a id="__codelineno-0-173" name="__codelineno-0-173"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-174" name="__codelineno-0-174"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-175" name="__codelineno-0-175"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-176" name="__codelineno-0-176"></a><span class="sd">        mismatches (int): Max number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-177" name="__codelineno-0-177"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-178" name="__codelineno-0-178"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-179" name="__codelineno-0-179"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-180" name="__codelineno-0-180"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-181" name="__codelineno-0-181"></a>
<a id="__codelineno-0-182" name="__codelineno-0-182"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-183" name="__codelineno-0-183"></a><span class="sd">        df (pd.DataFrame): dataframe with the following columns: description</span>
<a id="__codelineno-0-184" name="__codelineno-0-184"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-185" name="__codelineno-0-185"></a>
<a id="__codelineno-0-186" name="__codelineno-0-186"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating clonotypes count table</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-187" name="__codelineno-0-187"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-188" name="__codelineno-0-188"></a>
<a id="__codelineno-0-189" name="__codelineno-0-189"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-190" name="__codelineno-0-190"></a>
<a id="__codelineno-0-191" name="__codelineno-0-191"></a>    <span class="n">clonoset_dicts</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">,</span>
<a id="__codelineno-0-192" name="__codelineno-0-192"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-193" name="__codelineno-0-193"></a>
<a id="__codelineno-0-194" name="__codelineno-0-194"></a>    <span class="n">clonotypes_by_cluster</span> <span class="o">=</span> <span class="n">convert_clusters_to_clonotype_list</span><span class="p">(</span><span class="n">clusters_list</span><span class="p">,</span> <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">)</span>
<a id="__codelineno-0-195" name="__codelineno-0-195"></a>
<a id="__codelineno-0-196" name="__codelineno-0-196"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-197" name="__codelineno-0-197"></a>    <span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">clonoset_dicts</span><span class="p">:</span>
<a id="__codelineno-0-198" name="__codelineno-0-198"></a>        <span class="n">task</span> <span class="o">=</span> <span class="p">[</span><span class="n">clonotypes_by_cluster</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">clonoset_dicts</span><span class="p">[</span><span class="n">sample_id</span><span class="p">],</span> <span class="n">mismatches</span><span class="p">]</span>
<a id="__codelineno-0-199" name="__codelineno-0-199"></a>        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
<a id="__codelineno-0-200" name="__codelineno-0-200"></a>
<a id="__codelineno-0-201" name="__codelineno-0-201"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">count_table_by_cluster_mp</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Counting cluster presence&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;clonosets&quot;</span><span class="p">)</span>
<a id="__codelineno-0-202" name="__codelineno-0-202"></a>    <span class="n">result_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-203" name="__codelineno-0-203"></a>    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
<a id="__codelineno-0-204" name="__codelineno-0-204"></a>        <span class="n">result_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<a id="__codelineno-0-205" name="__codelineno-0-205"></a>    <span class="n">count_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_dict</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span><span class="s2">&quot;feature_id&quot;</span><span class="p">})</span>
<a id="__codelineno-0-206" name="__codelineno-0-206"></a>    <span class="n">count_table</span><span class="p">[</span><span class="s2">&quot;feature_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">count_table</span><span class="p">[</span><span class="s2">&quot;feature_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;cluster_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-207" name="__codelineno-0-207"></a>
<a id="__codelineno-0-208" name="__codelineno-0-208"></a>    <span class="k">return</span> <span class="n">count_table</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.count_table_with_custom_clonotypes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">count_table_with_custom_clonotypes</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clones_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">strict_presence</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">



<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clones_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>a dataframe with clonotypes to look for in <code>clonosets_df</code>. Should have the following columns:</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-731">731</a></span>
<span class="normal"><a href="#__codelineno-0-732">732</a></span>
<span class="normal"><a href="#__codelineno-0-733">733</a></span>
<span class="normal"><a href="#__codelineno-0-734">734</a></span>
<span class="normal"><a href="#__codelineno-0-735">735</a></span>
<span class="normal"><a href="#__codelineno-0-736">736</a></span>
<span class="normal"><a href="#__codelineno-0-737">737</a></span>
<span class="normal"><a href="#__codelineno-0-738">738</a></span>
<span class="normal"><a href="#__codelineno-0-739">739</a></span>
<span class="normal"><a href="#__codelineno-0-740">740</a></span>
<span class="normal"><a href="#__codelineno-0-741">741</a></span>
<span class="normal"><a href="#__codelineno-0-742">742</a></span>
<span class="normal"><a href="#__codelineno-0-743">743</a></span>
<span class="normal"><a href="#__codelineno-0-744">744</a></span>
<span class="normal"><a href="#__codelineno-0-745">745</a></span>
<span class="normal"><a href="#__codelineno-0-746">746</a></span>
<span class="normal"><a href="#__codelineno-0-747">747</a></span>
<span class="normal"><a href="#__codelineno-0-748">748</a></span>
<span class="normal"><a href="#__codelineno-0-749">749</a></span>
<span class="normal"><a href="#__codelineno-0-750">750</a></span>
<span class="normal"><a href="#__codelineno-0-751">751</a></span>
<span class="normal"><a href="#__codelineno-0-752">752</a></span>
<span class="normal"><a href="#__codelineno-0-753">753</a></span>
<span class="normal"><a href="#__codelineno-0-754">754</a></span>
<span class="normal"><a href="#__codelineno-0-755">755</a></span>
<span class="normal"><a href="#__codelineno-0-756">756</a></span>
<span class="normal"><a href="#__codelineno-0-757">757</a></span>
<span class="normal"><a href="#__codelineno-0-758">758</a></span>
<span class="normal"><a href="#__codelineno-0-759">759</a></span>
<span class="normal"><a href="#__codelineno-0-760">760</a></span>
<span class="normal"><a href="#__codelineno-0-761">761</a></span>
<span class="normal"><a href="#__codelineno-0-762">762</a></span>
<span class="normal"><a href="#__codelineno-0-763">763</a></span>
<span class="normal"><a href="#__codelineno-0-764">764</a></span>
<span class="normal"><a href="#__codelineno-0-765">765</a></span>
<span class="normal"><a href="#__codelineno-0-766">766</a></span>
<span class="normal"><a href="#__codelineno-0-767">767</a></span>
<span class="normal"><a href="#__codelineno-0-768">768</a></span>
<span class="normal"><a href="#__codelineno-0-769">769</a></span>
<span class="normal"><a href="#__codelineno-0-770">770</a></span>
<span class="normal"><a href="#__codelineno-0-771">771</a></span>
<span class="normal"><a href="#__codelineno-0-772">772</a></span>
<span class="normal"><a href="#__codelineno-0-773">773</a></span>
<span class="normal"><a href="#__codelineno-0-774">774</a></span>
<span class="normal"><a href="#__codelineno-0-775">775</a></span>
<span class="normal"><a href="#__codelineno-0-776">776</a></span>
<span class="normal"><a href="#__codelineno-0-777">777</a></span>
<span class="normal"><a href="#__codelineno-0-778">778</a></span>
<span class="normal"><a href="#__codelineno-0-779">779</a></span>
<span class="normal"><a href="#__codelineno-0-780">780</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-731" name="__codelineno-0-731"></a><span class="k">def</span><span class="w"> </span><span class="nf">count_table_with_custom_clonotypes</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clones_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">strict_presence</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-732" name="__codelineno-0-732"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-733" name="__codelineno-0-733"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-734" name="__codelineno-0-734"></a><span class="sd">        clones_df (pd.DataFrame): a dataframe with clonotypes to look for in `clonosets_df`. Should have the following columns:</span>
<a id="__codelineno-0-735" name="__codelineno-0-735"></a><span class="sd">        `cdr3aa`, `cdr3nt` (at least one of these two), `v`, `j` (could be none of those). For instance, your input might have only `cdr3aa` column.</span>
<a id="__codelineno-0-736" name="__codelineno-0-736"></a><span class="sd">        If `overlap_type` isn&#39;t specified, it is determined individually for each clonotype in `clones_df`. In this case, if a row corresponding to</span>
<a id="__codelineno-0-737" name="__codelineno-0-737"></a><span class="sd">        a particular clonotype has non-empty values in both `cdr3aa` and `cdr3nt` columns, `cdr3aa` is chosen.</span>
<a id="__codelineno-0-738" name="__codelineno-0-738"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-739" name="__codelineno-0-739"></a>
<a id="__codelineno-0-740" name="__codelineno-0-740"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Creating clonotypes count table</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-741" name="__codelineno-0-741"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-742" name="__codelineno-0-742"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-743" name="__codelineno-0-743"></a>    <span class="n">clonoset_dicts</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">,</span>
<a id="__codelineno-0-744" name="__codelineno-0-744"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-745" name="__codelineno-0-745"></a>
<a id="__codelineno-0-746" name="__codelineno-0-746"></a>    <span class="k">if</span> <span class="n">clones_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-747" name="__codelineno-0-747"></a>        <span class="c1"># creating dummy counts</span>
<a id="__codelineno-0-748" name="__codelineno-0-748"></a>        <span class="n">clones_df_copy</span> <span class="o">=</span> <span class="n">clones_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<a id="__codelineno-0-749" name="__codelineno-0-749"></a>        <span class="n">clones_df_copy</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">clones_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
<a id="__codelineno-0-750" name="__codelineno-0-750"></a>        <span class="n">clones_df_copy</span><span class="p">[</span><span class="s1">&#39;freq&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">clones_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
<a id="__codelineno-0-751" name="__codelineno-0-751"></a>        <span class="n">clonotype_list_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;clonotype_list&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
<a id="__codelineno-0-752" name="__codelineno-0-752"></a>
<a id="__codelineno-0-753" name="__codelineno-0-753"></a>        <span class="n">clonotype_list_dict</span><span class="p">[</span><span class="s1">&#39;clonotype_list&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">prepare_clonoset_for_intersection</span><span class="p">(</span><span class="n">clones_df_copy</span><span class="p">,</span> 
<a id="__codelineno-0-754" name="__codelineno-0-754"></a>                                                                                <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span>
<a id="__codelineno-0-755" name="__codelineno-0-755"></a>                                                                                <span class="n">len_vj_format</span><span class="o">=</span><span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>  
<a id="__codelineno-0-756" name="__codelineno-0-756"></a>
<a id="__codelineno-0-757" name="__codelineno-0-757"></a>        <span class="n">unique_clonotypes</span> <span class="o">=</span> <span class="n">find_unique_clonotypes_in_clonoset_dicts</span><span class="p">(</span><span class="n">clonotype_list_dict</span><span class="p">)</span>
<a id="__codelineno-0-758" name="__codelineno-0-758"></a>    <span class="k">else</span><span class="p">:</span> 
<a id="__codelineno-0-759" name="__codelineno-0-759"></a>        <span class="n">unique_clonotypes</span> <span class="o">=</span> <span class="n">find_unique_clonotypes_in_clonoset_dicts</span><span class="p">(</span><span class="n">clonoset_dicts</span><span class="p">)</span>
<a id="__codelineno-0-760" name="__codelineno-0-760"></a>
<a id="__codelineno-0-761" name="__codelineno-0-761"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-762" name="__codelineno-0-762"></a>    <span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">clonoset_dicts</span><span class="p">:</span>
<a id="__codelineno-0-763" name="__codelineno-0-763"></a>        <span class="n">task</span> <span class="o">=</span> <span class="p">[</span><span class="n">unique_clonotypes</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">clonoset_dicts</span><span class="p">[</span><span class="n">sample_id</span><span class="p">],</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">strict_presence</span><span class="p">]</span>
<a id="__codelineno-0-764" name="__codelineno-0-764"></a>        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
<a id="__codelineno-0-765" name="__codelineno-0-765"></a>
<a id="__codelineno-0-766" name="__codelineno-0-766"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">count_table_mp</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Counting features&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;clonosets&quot;</span><span class="p">)</span>
<a id="__codelineno-0-767" name="__codelineno-0-767"></a>    <span class="n">result_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<a id="__codelineno-0-768" name="__codelineno-0-768"></a>    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
<a id="__codelineno-0-769" name="__codelineno-0-769"></a>        <span class="n">result_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<a id="__codelineno-0-770" name="__codelineno-0-770"></a>    <span class="n">count_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_dict</span><span class="p">)</span>
<a id="__codelineno-0-771" name="__codelineno-0-771"></a>    <span class="k">if</span> <span class="n">aa</span><span class="p">:</span>
<a id="__codelineno-0-772" name="__codelineno-0-772"></a>        <span class="n">count_table</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;cdr3aa&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">ct</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">unique_clonotypes</span><span class="p">])</span>
<a id="__codelineno-0-773" name="__codelineno-0-773"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-774" name="__codelineno-0-774"></a>        <span class="n">count_table</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;cdr3nt&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">ct</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">unique_clonotypes</span><span class="p">])</span>
<a id="__codelineno-0-775" name="__codelineno-0-775"></a>    <span class="k">if</span> <span class="n">check_v</span><span class="p">:</span>
<a id="__codelineno-0-776" name="__codelineno-0-776"></a>        <span class="n">count_table</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">ct</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">unique_clonotypes</span><span class="p">])</span>
<a id="__codelineno-0-777" name="__codelineno-0-777"></a>    <span class="k">if</span> <span class="n">check_j</span><span class="p">:</span>
<a id="__codelineno-0-778" name="__codelineno-0-778"></a>        <span class="n">count_table</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;j&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">ct</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">unique_clonotypes</span><span class="p">])</span>
<a id="__codelineno-0-779" name="__codelineno-0-779"></a>    <span class="n">count_table</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">unique_clonotypes</span>
<a id="__codelineno-0-780" name="__codelineno-0-780"></a>    <span class="k">return</span> <span class="n">count_table</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.find_intersecting_clonotypes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">find_intersecting_clonotypes</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;F2&#39;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculating overlap distances between multiple repseq samples using F2 of F metrics
The result of this function may be used for heatmap+clusterization of samples or for MDS plots</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cl_filter</code>
            </td>
            <td>
                  <code><span title="repseq.clone_filter.Filter">Filter</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>clonoset filter - object from <code>clone_filter.py</code> module. It is applied to clonosets in <code>clonosets_df</code>.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>overlap_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mismatches</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Max number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>by_umi</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>set =True for MiXCR4 clonosets to select count/frequency of clonotypes 
in UMI's if they exist in implemented protocol</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>metric</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values - <code>F</code>, <code>F2</code> or <code>C</code>. Default <code>F2</code>. <code>F2</code> - sum of sqrt of product of 
similar clonotype frequencies in two clonosets. <code>F</code> - sqrt of the sum of frequency products.
<code>C</code> - total frequency of clonotypes in <code>sample1</code>, that are similar to clonotypes in <code>sample2</code></p>
              </div>
            </td>
            <td>
                  <code>&#39;F2&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df2</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>clonosets_df2</code> is None (default), samples within <code>clonosets_df</code> are compared with each other; 
Otherwise, the comparison is performed exclusively between samples from <code>clonosets_df</code> and <code>clonosets_df2</code>. </p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cl_filter2</code>
            </td>
            <td>
                  <code><span title="repseq.clone_filter.Filter">Filter</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>clonoset filter - object from <code>clone_filter.py</code> module. It is applied to clonosets in <code>clonosets_df2</code>. 
If there are samples with non-unique sample_ids between the two dataframes, both filters will be applied to those samples.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Important: similar clonotypes by <code>overlap_type</code> in one particular clonoset are NOT combined into one
and are treated as different clonotypes.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>clone</code>, <code>sample1_count</code>, <code>sample2_count</code>, <code>sample1</code>, <code>sample2</code>, <code>pair</code>. 
<code>clone</code> - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric
count columns contain freq/count of the clone in sample
pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-495">495</a></span>
<span class="normal"><a href="#__codelineno-0-496">496</a></span>
<span class="normal"><a href="#__codelineno-0-497">497</a></span>
<span class="normal"><a href="#__codelineno-0-498">498</a></span>
<span class="normal"><a href="#__codelineno-0-499">499</a></span>
<span class="normal"><a href="#__codelineno-0-500">500</a></span>
<span class="normal"><a href="#__codelineno-0-501">501</a></span>
<span class="normal"><a href="#__codelineno-0-502">502</a></span>
<span class="normal"><a href="#__codelineno-0-503">503</a></span>
<span class="normal"><a href="#__codelineno-0-504">504</a></span>
<span class="normal"><a href="#__codelineno-0-505">505</a></span>
<span class="normal"><a href="#__codelineno-0-506">506</a></span>
<span class="normal"><a href="#__codelineno-0-507">507</a></span>
<span class="normal"><a href="#__codelineno-0-508">508</a></span>
<span class="normal"><a href="#__codelineno-0-509">509</a></span>
<span class="normal"><a href="#__codelineno-0-510">510</a></span>
<span class="normal"><a href="#__codelineno-0-511">511</a></span>
<span class="normal"><a href="#__codelineno-0-512">512</a></span>
<span class="normal"><a href="#__codelineno-0-513">513</a></span>
<span class="normal"><a href="#__codelineno-0-514">514</a></span>
<span class="normal"><a href="#__codelineno-0-515">515</a></span>
<span class="normal"><a href="#__codelineno-0-516">516</a></span>
<span class="normal"><a href="#__codelineno-0-517">517</a></span>
<span class="normal"><a href="#__codelineno-0-518">518</a></span>
<span class="normal"><a href="#__codelineno-0-519">519</a></span>
<span class="normal"><a href="#__codelineno-0-520">520</a></span>
<span class="normal"><a href="#__codelineno-0-521">521</a></span>
<span class="normal"><a href="#__codelineno-0-522">522</a></span>
<span class="normal"><a href="#__codelineno-0-523">523</a></span>
<span class="normal"><a href="#__codelineno-0-524">524</a></span>
<span class="normal"><a href="#__codelineno-0-525">525</a></span>
<span class="normal"><a href="#__codelineno-0-526">526</a></span>
<span class="normal"><a href="#__codelineno-0-527">527</a></span>
<span class="normal"><a href="#__codelineno-0-528">528</a></span>
<span class="normal"><a href="#__codelineno-0-529">529</a></span>
<span class="normal"><a href="#__codelineno-0-530">530</a></span>
<span class="normal"><a href="#__codelineno-0-531">531</a></span>
<span class="normal"><a href="#__codelineno-0-532">532</a></span>
<span class="normal"><a href="#__codelineno-0-533">533</a></span>
<span class="normal"><a href="#__codelineno-0-534">534</a></span>
<span class="normal"><a href="#__codelineno-0-535">535</a></span>
<span class="normal"><a href="#__codelineno-0-536">536</a></span>
<span class="normal"><a href="#__codelineno-0-537">537</a></span>
<span class="normal"><a href="#__codelineno-0-538">538</a></span>
<span class="normal"><a href="#__codelineno-0-539">539</a></span>
<span class="normal"><a href="#__codelineno-0-540">540</a></span>
<span class="normal"><a href="#__codelineno-0-541">541</a></span>
<span class="normal"><a href="#__codelineno-0-542">542</a></span>
<span class="normal"><a href="#__codelineno-0-543">543</a></span>
<span class="normal"><a href="#__codelineno-0-544">544</a></span>
<span class="normal"><a href="#__codelineno-0-545">545</a></span>
<span class="normal"><a href="#__codelineno-0-546">546</a></span>
<span class="normal"><a href="#__codelineno-0-547">547</a></span>
<span class="normal"><a href="#__codelineno-0-548">548</a></span>
<span class="normal"><a href="#__codelineno-0-549">549</a></span>
<span class="normal"><a href="#__codelineno-0-550">550</a></span>
<span class="normal"><a href="#__codelineno-0-551">551</a></span>
<span class="normal"><a href="#__codelineno-0-552">552</a></span>
<span class="normal"><a href="#__codelineno-0-553">553</a></span>
<span class="normal"><a href="#__codelineno-0-554">554</a></span>
<span class="normal"><a href="#__codelineno-0-555">555</a></span>
<span class="normal"><a href="#__codelineno-0-556">556</a></span>
<span class="normal"><a href="#__codelineno-0-557">557</a></span>
<span class="normal"><a href="#__codelineno-0-558">558</a></span>
<span class="normal"><a href="#__codelineno-0-559">559</a></span>
<span class="normal"><a href="#__codelineno-0-560">560</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-495" name="__codelineno-0-495"></a><span class="k">def</span><span class="w"> </span><span class="nf">find_intersecting_clonotypes</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-496" name="__codelineno-0-496"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-497" name="__codelineno-0-497"></a><span class="sd">    Calculating overlap distances between multiple repseq samples using F2 of F metrics</span>
<a id="__codelineno-0-498" name="__codelineno-0-498"></a><span class="sd">    The result of this function may be used for heatmap+clusterization of samples or for MDS plots</span>
<a id="__codelineno-0-499" name="__codelineno-0-499"></a>
<a id="__codelineno-0-500" name="__codelineno-0-500"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-501" name="__codelineno-0-501"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-502" name="__codelineno-0-502"></a><span class="sd">            filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-503" name="__codelineno-0-503"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-504" name="__codelineno-0-504"></a><span class="sd">        cl_filter (Filter): clonoset filter - object from `clone_filter.py` module. It is applied to clonosets in `clonosets_df`.</span>
<a id="__codelineno-0-505" name="__codelineno-0-505"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-506" name="__codelineno-0-506"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-507" name="__codelineno-0-507"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-508" name="__codelineno-0-508"></a><span class="sd">        mismatches (int): Max number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-509" name="__codelineno-0-509"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-510" name="__codelineno-0-510"></a><span class="sd">        by_umi (bool): set =True for MiXCR4 clonosets to select count/frequency of clonotypes </span>
<a id="__codelineno-0-511" name="__codelineno-0-511"></a><span class="sd">            in UMI&#39;s if they exist in implemented protocol</span>
<a id="__codelineno-0-512" name="__codelineno-0-512"></a><span class="sd">        metric (str): possible values - `F`, `F2` or `C`. Default `F2`. `F2` - sum of sqrt of product of </span>
<a id="__codelineno-0-513" name="__codelineno-0-513"></a><span class="sd">            similar clonotype frequencies in two clonosets. `F` - sqrt of the sum of frequency products.</span>
<a id="__codelineno-0-514" name="__codelineno-0-514"></a><span class="sd">            `C` - total frequency of clonotypes in `sample1`, that are similar to clonotypes in `sample2`</span>
<a id="__codelineno-0-515" name="__codelineno-0-515"></a><span class="sd">        clonosets_df2 (pd.DataFrame): If `clonosets_df2` is None (default), samples within `clonosets_df` are compared with each other; </span>
<a id="__codelineno-0-516" name="__codelineno-0-516"></a><span class="sd">            Otherwise, the comparison is performed exclusively between samples from `clonosets_df` and `clonosets_df2`. </span>
<a id="__codelineno-0-517" name="__codelineno-0-517"></a><span class="sd">        cl_filter2 (Filter): clonoset filter - object from `clone_filter.py` module. It is applied to clonosets in `clonosets_df2`. </span>
<a id="__codelineno-0-518" name="__codelineno-0-518"></a><span class="sd">            If there are samples with non-unique sample_ids between the two dataframes, both filters will be applied to those samples.</span>
<a id="__codelineno-0-519" name="__codelineno-0-519"></a>
<a id="__codelineno-0-520" name="__codelineno-0-520"></a>
<a id="__codelineno-0-521" name="__codelineno-0-521"></a><span class="sd">    Important: similar clonotypes by `overlap_type` in one particular clonoset are NOT combined into one</span>
<a id="__codelineno-0-522" name="__codelineno-0-522"></a><span class="sd">    and are treated as different clonotypes.</span>
<a id="__codelineno-0-523" name="__codelineno-0-523"></a>
<a id="__codelineno-0-524" name="__codelineno-0-524"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-525" name="__codelineno-0-525"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `clone`, `sample1_count`, `sample2_count`, `sample1`, `sample2`, `pair`. </span>
<a id="__codelineno-0-526" name="__codelineno-0-526"></a><span class="sd">            `clone` - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric</span>
<a id="__codelineno-0-527" name="__codelineno-0-527"></a><span class="sd">            count columns contain freq/count of the clone in sample</span>
<a id="__codelineno-0-528" name="__codelineno-0-528"></a><span class="sd">            pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</span>
<a id="__codelineno-0-529" name="__codelineno-0-529"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-530" name="__codelineno-0-530"></a>
<a id="__codelineno-0-531" name="__codelineno-0-531"></a>
<a id="__codelineno-0-532" name="__codelineno-0-532"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Intersecting clones in clonosets</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-533" name="__codelineno-0-533"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-534" name="__codelineno-0-534"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-535" name="__codelineno-0-535"></a>
<a id="__codelineno-0-536" name="__codelineno-0-536"></a>    <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span> <span class="o">=</span> <span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span>
<a id="__codelineno-0-537" name="__codelineno-0-537"></a>                                                                                                                        <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-538" name="__codelineno-0-538"></a>                                                                                                                        <span class="n">overlap_type</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="ow">not</span> <span class="nb">bool</span><span class="p">(</span><span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-539" name="__codelineno-0-539"></a>
<a id="__codelineno-0-540" name="__codelineno-0-540"></a>    <span class="c1"># generating a set of tasks</span>
<a id="__codelineno-0-541" name="__codelineno-0-541"></a>
<a id="__codelineno-0-542" name="__codelineno-0-542"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-543" name="__codelineno-0-543"></a>
<a id="__codelineno-0-544" name="__codelineno-0-544"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-545" name="__codelineno-0-545"></a>        <span class="k">for</span> <span class="n">sample1</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
<a id="__codelineno-0-546" name="__codelineno-0-546"></a>            <span class="k">for</span> <span class="n">sample2</span> <span class="ow">in</span> <span class="n">sample_list2</span><span class="p">:</span>
<a id="__codelineno-0-547" name="__codelineno-0-547"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">))</span>        
<a id="__codelineno-0-548" name="__codelineno-0-548"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-549" name="__codelineno-0-549"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-550" name="__codelineno-0-550"></a>            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-551" name="__codelineno-0-551"></a>                <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-552" name="__codelineno-0-552"></a>                <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
<a id="__codelineno-0-553" name="__codelineno-0-553"></a>                <span class="k">if</span> <span class="n">sample1</span> <span class="o">!=</span> <span class="n">sample2</span><span class="p">:</span>
<a id="__codelineno-0-554" name="__codelineno-0-554"></a>                    <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">))</span>
<a id="__codelineno-0-555" name="__codelineno-0-555"></a>
<a id="__codelineno-0-556" name="__codelineno-0-556"></a>
<a id="__codelineno-0-557" name="__codelineno-0-557"></a>    <span class="c1"># run calculation in parallel</span>
<a id="__codelineno-0-558" name="__codelineno-0-558"></a>    <span class="n">result_list</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">find_overlapping_clones_in_two_clone_dicts</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Intersecting clonosets&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;pairs&quot;</span><span class="p">)</span>
<a id="__codelineno-0-559" name="__codelineno-0-559"></a>
<a id="__codelineno-0-560" name="__codelineno-0-560"></a>    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">result_list</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.intersect_clones_in_samples_batch" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">intersect_clones_in_samples_batch</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculating frequencies of intersecting clonotypes between multiple repseq samples.
The result of this function may be used for scatterplots of frequencies/counts of 
overlapping clonotypes</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
<code>filename</code> - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>overlap_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>by_umi</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>set <code>=True</code> for MiXCR4 clonosets to select count/frequency of clonotypes 
in UMI's if they exist in implemented protocol</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>by_freq</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>default is <code>True</code> - this means that the intersect metric is frequency of clonotype, 
but not its count</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>only_functional</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Important: when using particular overlap type, similar clonotypes in one particular clonoset are
combined into one with summation of counts/frequencies.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>clone</code>, <code>sample1_count</code>, <code>sample2_count</code>, <code>sample1</code>, <code>sample2</code>, <code>pair</code>
clone - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric
count columns contain freq/count of the clone in sample
pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-20">20</a></span>
<span class="normal"><a href="#__codelineno-0-21">21</a></span>
<span class="normal"><a href="#__codelineno-0-22">22</a></span>
<span class="normal"><a href="#__codelineno-0-23">23</a></span>
<span class="normal"><a href="#__codelineno-0-24">24</a></span>
<span class="normal"><a href="#__codelineno-0-25">25</a></span>
<span class="normal"><a href="#__codelineno-0-26">26</a></span>
<span class="normal"><a href="#__codelineno-0-27">27</a></span>
<span class="normal"><a href="#__codelineno-0-28">28</a></span>
<span class="normal"><a href="#__codelineno-0-29">29</a></span>
<span class="normal"><a href="#__codelineno-0-30">30</a></span>
<span class="normal"><a href="#__codelineno-0-31">31</a></span>
<span class="normal"><a href="#__codelineno-0-32">32</a></span>
<span class="normal"><a href="#__codelineno-0-33">33</a></span>
<span class="normal"><a href="#__codelineno-0-34">34</a></span>
<span class="normal"><a href="#__codelineno-0-35">35</a></span>
<span class="normal"><a href="#__codelineno-0-36">36</a></span>
<span class="normal"><a href="#__codelineno-0-37">37</a></span>
<span class="normal"><a href="#__codelineno-0-38">38</a></span>
<span class="normal"><a href="#__codelineno-0-39">39</a></span>
<span class="normal"><a href="#__codelineno-0-40">40</a></span>
<span class="normal"><a href="#__codelineno-0-41">41</a></span>
<span class="normal"><a href="#__codelineno-0-42">42</a></span>
<span class="normal"><a href="#__codelineno-0-43">43</a></span>
<span class="normal"><a href="#__codelineno-0-44">44</a></span>
<span class="normal"><a href="#__codelineno-0-45">45</a></span>
<span class="normal"><a href="#__codelineno-0-46">46</a></span>
<span class="normal"><a href="#__codelineno-0-47">47</a></span>
<span class="normal"><a href="#__codelineno-0-48">48</a></span>
<span class="normal"><a href="#__codelineno-0-49">49</a></span>
<span class="normal"><a href="#__codelineno-0-50">50</a></span>
<span class="normal"><a href="#__codelineno-0-51">51</a></span>
<span class="normal"><a href="#__codelineno-0-52">52</a></span>
<span class="normal"><a href="#__codelineno-0-53">53</a></span>
<span class="normal"><a href="#__codelineno-0-54">54</a></span>
<span class="normal"><a href="#__codelineno-0-55">55</a></span>
<span class="normal"><a href="#__codelineno-0-56">56</a></span>
<span class="normal"><a href="#__codelineno-0-57">57</a></span>
<span class="normal"><a href="#__codelineno-0-58">58</a></span>
<span class="normal"><a href="#__codelineno-0-59">59</a></span>
<span class="normal"><a href="#__codelineno-0-60">60</a></span>
<span class="normal"><a href="#__codelineno-0-61">61</a></span>
<span class="normal"><a href="#__codelineno-0-62">62</a></span>
<span class="normal"><a href="#__codelineno-0-63">63</a></span>
<span class="normal"><a href="#__codelineno-0-64">64</a></span>
<span class="normal"><a href="#__codelineno-0-65">65</a></span>
<span class="normal"><a href="#__codelineno-0-66">66</a></span>
<span class="normal"><a href="#__codelineno-0-67">67</a></span>
<span class="normal"><a href="#__codelineno-0-68">68</a></span>
<span class="normal"><a href="#__codelineno-0-69">69</a></span>
<span class="normal"><a href="#__codelineno-0-70">70</a></span>
<span class="normal"><a href="#__codelineno-0-71">71</a></span>
<span class="normal"><a href="#__codelineno-0-72">72</a></span>
<span class="normal"><a href="#__codelineno-0-73">73</a></span>
<span class="normal"><a href="#__codelineno-0-74">74</a></span>
<span class="normal"><a href="#__codelineno-0-75">75</a></span>
<span class="normal"><a href="#__codelineno-0-76">76</a></span>
<span class="normal"><a href="#__codelineno-0-77">77</a></span>
<span class="normal"><a href="#__codelineno-0-78">78</a></span>
<span class="normal"><a href="#__codelineno-0-79">79</a></span>
<span class="normal"><a href="#__codelineno-0-80">80</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-20" name="__codelineno-0-20"></a><span class="k">def</span><span class="w"> </span><span class="nf">intersect_clones_in_samples_batch</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-21" name="__codelineno-0-21"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-22" name="__codelineno-0-22"></a><span class="sd">    Calculating frequencies of intersecting clonotypes between multiple repseq samples.</span>
<a id="__codelineno-0-23" name="__codelineno-0-23"></a><span class="sd">    The result of this function may be used for scatterplots of frequencies/counts of </span>
<a id="__codelineno-0-24" name="__codelineno-0-24"></a><span class="sd">    overlapping clonotypes</span>
<a id="__codelineno-0-25" name="__codelineno-0-25"></a>
<a id="__codelineno-0-26" name="__codelineno-0-26"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-27" name="__codelineno-0-27"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-28" name="__codelineno-0-28"></a><span class="sd">            `filename` - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-29" name="__codelineno-0-29"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-30" name="__codelineno-0-30"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-31" name="__codelineno-0-31"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-32" name="__codelineno-0-32"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-33" name="__codelineno-0-33"></a><span class="sd">        by_umi (bool): set `=True` for MiXCR4 clonosets to select count/frequency of clonotypes </span>
<a id="__codelineno-0-34" name="__codelineno-0-34"></a><span class="sd">            in UMI&#39;s if they exist in implemented protocol</span>
<a id="__codelineno-0-35" name="__codelineno-0-35"></a><span class="sd">        by_freq (bool): default is `True` - this means that the intersect metric is frequency of clonotype, </span>
<a id="__codelineno-0-36" name="__codelineno-0-36"></a><span class="sd">            but not its count</span>
<a id="__codelineno-0-37" name="__codelineno-0-37"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-38" name="__codelineno-0-38"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-39" name="__codelineno-0-39"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-40" name="__codelineno-0-40"></a>
<a id="__codelineno-0-41" name="__codelineno-0-41"></a><span class="sd">    Important: when using particular overlap type, similar clonotypes in one particular clonoset are</span>
<a id="__codelineno-0-42" name="__codelineno-0-42"></a><span class="sd">    combined into one with summation of counts/frequencies.</span>
<a id="__codelineno-0-43" name="__codelineno-0-43"></a>
<a id="__codelineno-0-44" name="__codelineno-0-44"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-45" name="__codelineno-0-45"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `clone`, `sample1_count`, `sample2_count`, `sample1`, `sample2`, `pair`</span>
<a id="__codelineno-0-46" name="__codelineno-0-46"></a><span class="sd">            clone - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric</span>
<a id="__codelineno-0-47" name="__codelineno-0-47"></a><span class="sd">            count columns contain freq/count of the clone in sample</span>
<a id="__codelineno-0-48" name="__codelineno-0-48"></a><span class="sd">            pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</span>
<a id="__codelineno-0-49" name="__codelineno-0-49"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-50" name="__codelineno-0-50"></a>
<a id="__codelineno-0-51" name="__codelineno-0-51"></a>
<a id="__codelineno-0-52" name="__codelineno-0-52"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Intersecting clones in clonosets</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-53" name="__codelineno-0-53"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-54" name="__codelineno-0-54"></a>
<a id="__codelineno-0-55" name="__codelineno-0-55"></a>    <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span> <span class="o">=</span> <span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span>
<a id="__codelineno-0-56" name="__codelineno-0-56"></a>                                                                                                                        <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-57" name="__codelineno-0-57"></a>                                                                                                                        <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span>
<a id="__codelineno-0-58" name="__codelineno-0-58"></a>                                                                                                                        <span class="n">strict</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-59" name="__codelineno-0-59"></a>    <span class="c1"># generating a set of tasks</span>
<a id="__codelineno-0-60" name="__codelineno-0-60"></a>
<a id="__codelineno-0-61" name="__codelineno-0-61"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-62" name="__codelineno-0-62"></a>
<a id="__codelineno-0-63" name="__codelineno-0-63"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-64" name="__codelineno-0-64"></a>        <span class="k">for</span> <span class="n">sample1</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
<a id="__codelineno-0-65" name="__codelineno-0-65"></a>            <span class="k">for</span> <span class="n">sample2</span> <span class="ow">in</span> <span class="n">sample_list2</span><span class="p">:</span>
<a id="__codelineno-0-66" name="__codelineno-0-66"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">))</span>
<a id="__codelineno-0-67" name="__codelineno-0-67"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-68" name="__codelineno-0-68"></a>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-69" name="__codelineno-0-69"></a>            <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-70" name="__codelineno-0-70"></a>            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="o">-</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-71" name="__codelineno-0-71"></a>                <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-72" name="__codelineno-0-72"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">))</span>
<a id="__codelineno-0-73" name="__codelineno-0-73"></a>
<a id="__codelineno-0-74" name="__codelineno-0-74"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">intersect_two_clone_dicts</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Intersecting clonosets&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;pairs&quot;</span><span class="p">)</span>
<a id="__codelineno-0-75" name="__codelineno-0-75"></a>
<a id="__codelineno-0-76" name="__codelineno-0-76"></a>    <span class="c1"># df = pd.concat(results).index.set_names()</span>
<a id="__codelineno-0-77" name="__codelineno-0-77"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">results</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-78" name="__codelineno-0-78"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">split_tuple_clone_column</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-79" name="__codelineno-0-79"></a>
<a id="__codelineno-0-80" name="__codelineno-0-80"></a>    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.overlap_distances" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">overlap_distances</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaV&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;F2&#39;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Calculating overlap distances between multiple repseq samples using F2 of F metrics
The result of this function may be used for heatmap+clusterization of samples or for MDS plots</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>contains three columns - <code>sample_id</code> and <code>filename</code> columns,
filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
sample_id's should be all unique in this DF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>overlap_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
to decide if clonotypes are equal</p>
              </div>
            </td>
            <td>
                  <code>&#39;aaV&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mismatches</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Max number of single-letter mismatches in clonotypes sequences 
for them to be treated similar, i.e. hamming distance.</p>
              </div>
            </td>
            <td>
                  <code>0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>by_umi</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>set =True for MiXCR4 clonosets to select count/frequency of clonotypes 
in UMI's if they exist in implemented protocol</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>metric</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>possible values - <code>F</code>, <code>F2</code> or <code>C</code>. Default <code>F2</code>. <code>F2</code> - sum of sqrt of product of 
similar clonotype frequencies in two clonosets. <code>F</code> - sqrt of the sum of frequency products.
<code>C</code> - total frequency of clonotypes in <code>sample1</code>, that are similar to clonotypes in <code>sample2</code></p>
              </div>
            </td>
            <td>
                  <code>&#39;F2&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>only_functional</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>use only functional clonotypes (do not contain stop codons or
frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to
1 after filtering of non-functional clonotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<details class="important" open>
  <summary>similar clonotypes by <code>overlap_type</code> in one particular clonoset will be combined into one</summary>
  <p>clonotype with sum for count.</p>
</details>

    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>clone</code>, <code>sample1_count</code>, <code>sample2_count</code>, <code>sample1</code>, <code>sample2</code>, <code>pair</code>. 
<code>clone</code> - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric
count columns contain freq/count of the clone in sample
pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-403">403</a></span>
<span class="normal"><a href="#__codelineno-0-404">404</a></span>
<span class="normal"><a href="#__codelineno-0-405">405</a></span>
<span class="normal"><a href="#__codelineno-0-406">406</a></span>
<span class="normal"><a href="#__codelineno-0-407">407</a></span>
<span class="normal"><a href="#__codelineno-0-408">408</a></span>
<span class="normal"><a href="#__codelineno-0-409">409</a></span>
<span class="normal"><a href="#__codelineno-0-410">410</a></span>
<span class="normal"><a href="#__codelineno-0-411">411</a></span>
<span class="normal"><a href="#__codelineno-0-412">412</a></span>
<span class="normal"><a href="#__codelineno-0-413">413</a></span>
<span class="normal"><a href="#__codelineno-0-414">414</a></span>
<span class="normal"><a href="#__codelineno-0-415">415</a></span>
<span class="normal"><a href="#__codelineno-0-416">416</a></span>
<span class="normal"><a href="#__codelineno-0-417">417</a></span>
<span class="normal"><a href="#__codelineno-0-418">418</a></span>
<span class="normal"><a href="#__codelineno-0-419">419</a></span>
<span class="normal"><a href="#__codelineno-0-420">420</a></span>
<span class="normal"><a href="#__codelineno-0-421">421</a></span>
<span class="normal"><a href="#__codelineno-0-422">422</a></span>
<span class="normal"><a href="#__codelineno-0-423">423</a></span>
<span class="normal"><a href="#__codelineno-0-424">424</a></span>
<span class="normal"><a href="#__codelineno-0-425">425</a></span>
<span class="normal"><a href="#__codelineno-0-426">426</a></span>
<span class="normal"><a href="#__codelineno-0-427">427</a></span>
<span class="normal"><a href="#__codelineno-0-428">428</a></span>
<span class="normal"><a href="#__codelineno-0-429">429</a></span>
<span class="normal"><a href="#__codelineno-0-430">430</a></span>
<span class="normal"><a href="#__codelineno-0-431">431</a></span>
<span class="normal"><a href="#__codelineno-0-432">432</a></span>
<span class="normal"><a href="#__codelineno-0-433">433</a></span>
<span class="normal"><a href="#__codelineno-0-434">434</a></span>
<span class="normal"><a href="#__codelineno-0-435">435</a></span>
<span class="normal"><a href="#__codelineno-0-436">436</a></span>
<span class="normal"><a href="#__codelineno-0-437">437</a></span>
<span class="normal"><a href="#__codelineno-0-438">438</a></span>
<span class="normal"><a href="#__codelineno-0-439">439</a></span>
<span class="normal"><a href="#__codelineno-0-440">440</a></span>
<span class="normal"><a href="#__codelineno-0-441">441</a></span>
<span class="normal"><a href="#__codelineno-0-442">442</a></span>
<span class="normal"><a href="#__codelineno-0-443">443</a></span>
<span class="normal"><a href="#__codelineno-0-444">444</a></span>
<span class="normal"><a href="#__codelineno-0-445">445</a></span>
<span class="normal"><a href="#__codelineno-0-446">446</a></span>
<span class="normal"><a href="#__codelineno-0-447">447</a></span>
<span class="normal"><a href="#__codelineno-0-448">448</a></span>
<span class="normal"><a href="#__codelineno-0-449">449</a></span>
<span class="normal"><a href="#__codelineno-0-450">450</a></span>
<span class="normal"><a href="#__codelineno-0-451">451</a></span>
<span class="normal"><a href="#__codelineno-0-452">452</a></span>
<span class="normal"><a href="#__codelineno-0-453">453</a></span>
<span class="normal"><a href="#__codelineno-0-454">454</a></span>
<span class="normal"><a href="#__codelineno-0-455">455</a></span>
<span class="normal"><a href="#__codelineno-0-456">456</a></span>
<span class="normal"><a href="#__codelineno-0-457">457</a></span>
<span class="normal"><a href="#__codelineno-0-458">458</a></span>
<span class="normal"><a href="#__codelineno-0-459">459</a></span>
<span class="normal"><a href="#__codelineno-0-460">460</a></span>
<span class="normal"><a href="#__codelineno-0-461">461</a></span>
<span class="normal"><a href="#__codelineno-0-462">462</a></span>
<span class="normal"><a href="#__codelineno-0-463">463</a></span>
<span class="normal"><a href="#__codelineno-0-464">464</a></span>
<span class="normal"><a href="#__codelineno-0-465">465</a></span>
<span class="normal"><a href="#__codelineno-0-466">466</a></span>
<span class="normal"><a href="#__codelineno-0-467">467</a></span>
<span class="normal"><a href="#__codelineno-0-468">468</a></span>
<span class="normal"><a href="#__codelineno-0-469">469</a></span>
<span class="normal"><a href="#__codelineno-0-470">470</a></span>
<span class="normal"><a href="#__codelineno-0-471">471</a></span>
<span class="normal"><a href="#__codelineno-0-472">472</a></span>
<span class="normal"><a href="#__codelineno-0-473">473</a></span>
<span class="normal"><a href="#__codelineno-0-474">474</a></span>
<span class="normal"><a href="#__codelineno-0-475">475</a></span>
<span class="normal"><a href="#__codelineno-0-476">476</a></span>
<span class="normal"><a href="#__codelineno-0-477">477</a></span>
<span class="normal"><a href="#__codelineno-0-478">478</a></span>
<span class="normal"><a href="#__codelineno-0-479">479</a></span>
<span class="normal"><a href="#__codelineno-0-480">480</a></span>
<span class="normal"><a href="#__codelineno-0-481">481</a></span>
<span class="normal"><a href="#__codelineno-0-482">482</a></span>
<span class="normal"><a href="#__codelineno-0-483">483</a></span>
<span class="normal"><a href="#__codelineno-0-484">484</a></span>
<span class="normal"><a href="#__codelineno-0-485">485</a></span>
<span class="normal"><a href="#__codelineno-0-486">486</a></span>
<span class="normal"><a href="#__codelineno-0-487">487</a></span>
<span class="normal"><a href="#__codelineno-0-488">488</a></span>
<span class="normal"><a href="#__codelineno-0-489">489</a></span>
<span class="normal"><a href="#__codelineno-0-490">490</a></span>
<span class="normal"><a href="#__codelineno-0-491">491</a></span>
<span class="normal"><a href="#__codelineno-0-492">492</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-403" name="__codelineno-0-403"></a><span class="k">def</span><span class="w"> </span><span class="nf">overlap_distances</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaV&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<a id="__codelineno-0-404" name="__codelineno-0-404"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-405" name="__codelineno-0-405"></a><span class="sd">    Calculating overlap distances between multiple repseq samples using F2 of F metrics</span>
<a id="__codelineno-0-406" name="__codelineno-0-406"></a><span class="sd">    The result of this function may be used for heatmap+clusterization of samples or for MDS plots</span>
<a id="__codelineno-0-407" name="__codelineno-0-407"></a>
<a id="__codelineno-0-408" name="__codelineno-0-408"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-409" name="__codelineno-0-409"></a><span class="sd">        clonosets_df (pd.DataFrame): contains three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-410" name="__codelineno-0-410"></a><span class="sd">            filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-411" name="__codelineno-0-411"></a><span class="sd">            sample_id&#39;s should be all unique in this DF</span>
<a id="__codelineno-0-412" name="__codelineno-0-412"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-413" name="__codelineno-0-413"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-414" name="__codelineno-0-414"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-415" name="__codelineno-0-415"></a><span class="sd">        mismatches (int): Max number of single-letter mismatches in clonotypes sequences </span>
<a id="__codelineno-0-416" name="__codelineno-0-416"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-417" name="__codelineno-0-417"></a><span class="sd">        by_umi (bool): set =True for MiXCR4 clonosets to select count/frequency of clonotypes </span>
<a id="__codelineno-0-418" name="__codelineno-0-418"></a><span class="sd">            in UMI&#39;s if they exist in implemented protocol</span>
<a id="__codelineno-0-419" name="__codelineno-0-419"></a><span class="sd">        metric (str): possible values - `F`, `F2` or `C`. Default `F2`. `F2` - sum of sqrt of product of </span>
<a id="__codelineno-0-420" name="__codelineno-0-420"></a><span class="sd">            similar clonotype frequencies in two clonosets. `F` - sqrt of the sum of frequency products.</span>
<a id="__codelineno-0-421" name="__codelineno-0-421"></a><span class="sd">            `C` - total frequency of clonotypes in `sample1`, that are similar to clonotypes in `sample2`</span>
<a id="__codelineno-0-422" name="__codelineno-0-422"></a><span class="sd">        only_functional (bool): use only functional clonotypes (do not contain stop codons or</span>
<a id="__codelineno-0-423" name="__codelineno-0-423"></a><span class="sd">            frameshifts in CDR3 sequences: * or _ symbol in CDR3aa sequence). The frequences are recounted to</span>
<a id="__codelineno-0-424" name="__codelineno-0-424"></a><span class="sd">            1 after filtering of non-functional clonotypes</span>
<a id="__codelineno-0-425" name="__codelineno-0-425"></a>
<a id="__codelineno-0-426" name="__codelineno-0-426"></a><span class="sd">    Important: similar clonotypes by `overlap_type` in one particular clonoset will be combined into one</span>
<a id="__codelineno-0-427" name="__codelineno-0-427"></a><span class="sd">        clonotype with sum for count.</span>
<a id="__codelineno-0-428" name="__codelineno-0-428"></a>
<a id="__codelineno-0-429" name="__codelineno-0-429"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-430" name="__codelineno-0-430"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `clone`, `sample1_count`, `sample2_count`, `sample1`, `sample2`, `pair`. </span>
<a id="__codelineno-0-431" name="__codelineno-0-431"></a><span class="sd">            `clone` - is tuple, containing sequence (aa or nt), plus V or J if they are required by the metric</span>
<a id="__codelineno-0-432" name="__codelineno-0-432"></a><span class="sd">            count columns contain freq/count of the clone in sample</span>
<a id="__codelineno-0-433" name="__codelineno-0-433"></a><span class="sd">            pair column is made for easy separation of possibly huge DataFrame into overlapping pairs</span>
<a id="__codelineno-0-434" name="__codelineno-0-434"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-435" name="__codelineno-0-435"></a>
<a id="__codelineno-0-436" name="__codelineno-0-436"></a>
<a id="__codelineno-0-437" name="__codelineno-0-437"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Intersecting clones in clonosets</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-438" name="__codelineno-0-438"></a>    <span class="n">aa</span><span class="p">,</span> <span class="n">check_v</span><span class="p">,</span> <span class="n">check_j</span> <span class="o">=</span> <span class="n">overlap_type_to_flags</span><span class="p">(</span><span class="n">overlap_type</span><span class="p">)</span>
<a id="__codelineno-0-439" name="__codelineno-0-439"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-440" name="__codelineno-0-440"></a>
<a id="__codelineno-0-441" name="__codelineno-0-441"></a>    <span class="n">metric</span> <span class="o">=</span> <span class="n">metric</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<a id="__codelineno-0-442" name="__codelineno-0-442"></a>    <span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;BC&quot;</span><span class="p">,</span> <span class="s2">&quot;J&quot;</span><span class="p">,</span> <span class="s2">&quot;JSD&quot;</span><span class="p">]</span>
<a id="__codelineno-0-443" name="__codelineno-0-443"></a>    <span class="n">mismatch_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
<a id="__codelineno-0-444" name="__codelineno-0-444"></a>    <span class="n">non_symmetry_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">]</span>
<a id="__codelineno-0-445" name="__codelineno-0-445"></a>    <span class="n">frequency_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;F2&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
<a id="__codelineno-0-446" name="__codelineno-0-446"></a>
<a id="__codelineno-0-447" name="__codelineno-0-447"></a>
<a id="__codelineno-0-448" name="__codelineno-0-448"></a>    <span class="k">if</span> <span class="n">metric</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">metrics</span><span class="p">:</span>
<a id="__codelineno-0-449" name="__codelineno-0-449"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Metric </span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2"> is not supported. Possible values: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-450" name="__codelineno-0-450"></a>
<a id="__codelineno-0-451" name="__codelineno-0-451"></a>    <span class="k">if</span> <span class="n">mismatches</span> <span class="ow">and</span> <span class="n">metric</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mismatch_metrics</span><span class="p">:</span>
<a id="__codelineno-0-452" name="__codelineno-0-452"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Metric </span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2"> does not allow mismatches. Mismatches only possible for: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mismatch_metrics</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-453" name="__codelineno-0-453"></a>
<a id="__codelineno-0-454" name="__codelineno-0-454"></a>    <span class="n">by_freq</span> <span class="o">=</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">frequency_metrics</span>
<a id="__codelineno-0-455" name="__codelineno-0-455"></a>
<a id="__codelineno-0-456" name="__codelineno-0-456"></a>    <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span> <span class="o">=</span> <span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span>
<a id="__codelineno-0-457" name="__codelineno-0-457"></a>                                                                                                                        <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-458" name="__codelineno-0-458"></a>                                                                                                                        <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">)</span>
<a id="__codelineno-0-459" name="__codelineno-0-459"></a>
<a id="__codelineno-0-460" name="__codelineno-0-460"></a>    <span class="c1"># generating a set of tasks</span>
<a id="__codelineno-0-461" name="__codelineno-0-461"></a>
<a id="__codelineno-0-462" name="__codelineno-0-462"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-463" name="__codelineno-0-463"></a>
<a id="__codelineno-0-464" name="__codelineno-0-464"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-465" name="__codelineno-0-465"></a>        <span class="k">for</span> <span class="n">sample1</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
<a id="__codelineno-0-466" name="__codelineno-0-466"></a>            <span class="k">for</span> <span class="n">sample2</span> <span class="ow">in</span> <span class="n">sample_list2</span><span class="p">:</span>
<a id="__codelineno-0-467" name="__codelineno-0-467"></a>                <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>        
<a id="__codelineno-0-468" name="__codelineno-0-468"></a>    <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-469" name="__codelineno-0-469"></a>        <span class="k">if</span> <span class="n">metric</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">non_symmetry_metrics</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-470" name="__codelineno-0-470"></a>            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-471" name="__codelineno-0-471"></a>                <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-472" name="__codelineno-0-472"></a>                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="o">-</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-473" name="__codelineno-0-473"></a>                    <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
<a id="__codelineno-0-474" name="__codelineno-0-474"></a>                    <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-475" name="__codelineno-0-475"></a>                <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s2">&quot;F2&quot;</span><span class="p">:</span>
<a id="__codelineno-0-476" name="__codelineno-0-476"></a>                    <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample1</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-477" name="__codelineno-0-477"></a>        <span class="k">else</span><span class="p">:</span>
<a id="__codelineno-0-478" name="__codelineno-0-478"></a>            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-479" name="__codelineno-0-479"></a>                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">samples_total</span><span class="p">):</span>
<a id="__codelineno-0-480" name="__codelineno-0-480"></a>                    <span class="n">sample1</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<a id="__codelineno-0-481" name="__codelineno-0-481"></a>                    <span class="n">sample2</span> <span class="o">=</span> <span class="n">sample_list</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
<a id="__codelineno-0-482" name="__codelineno-0-482"></a>                    <span class="k">if</span> <span class="n">sample1</span> <span class="o">!=</span> <span class="n">sample2</span><span class="p">:</span>
<a id="__codelineno-0-483" name="__codelineno-0-483"></a>                        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">,</span> <span class="n">metric</span><span class="p">))</span>
<a id="__codelineno-0-484" name="__codelineno-0-484"></a>
<a id="__codelineno-0-485" name="__codelineno-0-485"></a>
<a id="__codelineno-0-486" name="__codelineno-0-486"></a>    <span class="c1"># run calculation in parallel</span>
<a id="__codelineno-0-487" name="__codelineno-0-487"></a>    <span class="n">result_list</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">overlap_metric_two_clone_dicts</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Intersecting clonosets&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;pairs&quot;</span><span class="p">)</span>
<a id="__codelineno-0-488" name="__codelineno-0-488"></a>
<a id="__codelineno-0-489" name="__codelineno-0-489"></a>    <span class="k">if</span> <span class="ow">not</span> <span class="n">two_dataframes</span> <span class="ow">and</span> <span class="n">metric</span> <span class="o">!=</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span>
<a id="__codelineno-0-490" name="__codelineno-0-490"></a>        <span class="n">result_list</span> <span class="o">=</span> <span class="n">result_list</span> <span class="o">+</span> <span class="p">[(</span><span class="n">result</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">result</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">result_list</span><span class="p">]</span>
<a id="__codelineno-0-491" name="__codelineno-0-491"></a>    <span class="n">overlap_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">result_list</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="s2">&quot;sample2&quot;</span><span class="p">,</span> <span class="n">metric</span><span class="o">.</span><span class="n">lower</span><span class="p">()])</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample2&quot;</span><span class="p">],</span> <span class="n">values</span><span class="o">=</span><span class="n">metric</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;sample1&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<a id="__codelineno-0-492" name="__codelineno-0-492"></a>    <span class="k">return</span> <span class="n">overlap_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.prepare_clonotypes_dfs_for_intersections" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span> <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span> <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">



<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clonosets_df2</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cl_filter</code>
            </td>
            <td>
                  <code><span title="repseq.clone_filter.Filter">Filter</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cl_filter2</code>
            </td>
            <td>
                  <code><span title="repseq.clone_filter.Filter">Filter</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>overlap_type</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>by_freq</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em>. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p><em>description</em></p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>clonoset_lists</code></td>            <td>
                  <code><span title="dict">dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict of </p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>samples_total</code></td>            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>two_dataframes</code></td>            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>sample_list</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>sample_list2</code></td>            <td>
                  <code><span title="list">list</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-641">641</a></span>
<span class="normal"><a href="#__codelineno-0-642">642</a></span>
<span class="normal"><a href="#__codelineno-0-643">643</a></span>
<span class="normal"><a href="#__codelineno-0-644">644</a></span>
<span class="normal"><a href="#__codelineno-0-645">645</a></span>
<span class="normal"><a href="#__codelineno-0-646">646</a></span>
<span class="normal"><a href="#__codelineno-0-647">647</a></span>
<span class="normal"><a href="#__codelineno-0-648">648</a></span>
<span class="normal"><a href="#__codelineno-0-649">649</a></span>
<span class="normal"><a href="#__codelineno-0-650">650</a></span>
<span class="normal"><a href="#__codelineno-0-651">651</a></span>
<span class="normal"><a href="#__codelineno-0-652">652</a></span>
<span class="normal"><a href="#__codelineno-0-653">653</a></span>
<span class="normal"><a href="#__codelineno-0-654">654</a></span>
<span class="normal"><a href="#__codelineno-0-655">655</a></span>
<span class="normal"><a href="#__codelineno-0-656">656</a></span>
<span class="normal"><a href="#__codelineno-0-657">657</a></span>
<span class="normal"><a href="#__codelineno-0-658">658</a></span>
<span class="normal"><a href="#__codelineno-0-659">659</a></span>
<span class="normal"><a href="#__codelineno-0-660">660</a></span>
<span class="normal"><a href="#__codelineno-0-661">661</a></span>
<span class="normal"><a href="#__codelineno-0-662">662</a></span>
<span class="normal"><a href="#__codelineno-0-663">663</a></span>
<span class="normal"><a href="#__codelineno-0-664">664</a></span>
<span class="normal"><a href="#__codelineno-0-665">665</a></span>
<span class="normal"><a href="#__codelineno-0-666">666</a></span>
<span class="normal"><a href="#__codelineno-0-667">667</a></span>
<span class="normal"><a href="#__codelineno-0-668">668</a></span>
<span class="normal"><a href="#__codelineno-0-669">669</a></span>
<span class="normal"><a href="#__codelineno-0-670">670</a></span>
<span class="normal"><a href="#__codelineno-0-671">671</a></span>
<span class="normal"><a href="#__codelineno-0-672">672</a></span>
<span class="normal"><a href="#__codelineno-0-673">673</a></span>
<span class="normal"><a href="#__codelineno-0-674">674</a></span>
<span class="normal"><a href="#__codelineno-0-675">675</a></span>
<span class="normal"><a href="#__codelineno-0-676">676</a></span>
<span class="normal"><a href="#__codelineno-0-677">677</a></span>
<span class="normal"><a href="#__codelineno-0-678">678</a></span>
<span class="normal"><a href="#__codelineno-0-679">679</a></span>
<span class="normal"><a href="#__codelineno-0-680">680</a></span>
<span class="normal"><a href="#__codelineno-0-681">681</a></span>
<span class="normal"><a href="#__codelineno-0-682">682</a></span>
<span class="normal"><a href="#__codelineno-0-683">683</a></span>
<span class="normal"><a href="#__codelineno-0-684">684</a></span>
<span class="normal"><a href="#__codelineno-0-685">685</a></span>
<span class="normal"><a href="#__codelineno-0-686">686</a></span>
<span class="normal"><a href="#__codelineno-0-687">687</a></span>
<span class="normal"><a href="#__codelineno-0-688">688</a></span>
<span class="normal"><a href="#__codelineno-0-689">689</a></span>
<span class="normal"><a href="#__codelineno-0-690">690</a></span>
<span class="normal"><a href="#__codelineno-0-691">691</a></span>
<span class="normal"><a href="#__codelineno-0-692">692</a></span>
<span class="normal"><a href="#__codelineno-0-693">693</a></span>
<span class="normal"><a href="#__codelineno-0-694">694</a></span>
<span class="normal"><a href="#__codelineno-0-695">695</a></span>
<span class="normal"><a href="#__codelineno-0-696">696</a></span>
<span class="normal"><a href="#__codelineno-0-697">697</a></span>
<span class="normal"><a href="#__codelineno-0-698">698</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-641" name="__codelineno-0-641"></a><span class="k">def</span><span class="w"> </span><span class="nf">prepare_clonotypes_dfs_for_intersections</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">,</span> <span class="n">clonosets_df2</span><span class="p">,</span> <span class="n">cl_filter</span><span class="p">,</span> <span class="n">cl_filter2</span><span class="p">,</span> <span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<a id="__codelineno-0-642" name="__codelineno-0-642"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-643" name="__codelineno-0-643"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-644" name="__codelineno-0-644"></a><span class="sd">        clonosets_df (pd.DataFrame): _description_</span>
<a id="__codelineno-0-645" name="__codelineno-0-645"></a><span class="sd">        clonosets_df2 (pd.DataFrame): _description_</span>
<a id="__codelineno-0-646" name="__codelineno-0-646"></a><span class="sd">        cl_filter (Filter): _description_</span>
<a id="__codelineno-0-647" name="__codelineno-0-647"></a><span class="sd">        cl_filter2 (Filter): _description_</span>
<a id="__codelineno-0-648" name="__codelineno-0-648"></a><span class="sd">        overlap_type (str): _description_</span>
<a id="__codelineno-0-649" name="__codelineno-0-649"></a><span class="sd">        by_freq (bool, optional): _description_. Defaults to True.</span>
<a id="__codelineno-0-650" name="__codelineno-0-650"></a>
<a id="__codelineno-0-651" name="__codelineno-0-651"></a><span class="sd">    Raises:</span>
<a id="__codelineno-0-652" name="__codelineno-0-652"></a><span class="sd">        ValueError: _description_</span>
<a id="__codelineno-0-653" name="__codelineno-0-653"></a><span class="sd">        ValueError: _description_</span>
<a id="__codelineno-0-654" name="__codelineno-0-654"></a>
<a id="__codelineno-0-655" name="__codelineno-0-655"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-656" name="__codelineno-0-656"></a><span class="sd">        clonoset_lists (dict): dict of </span>
<a id="__codelineno-0-657" name="__codelineno-0-657"></a><span class="sd">        samples_total (int): </span>
<a id="__codelineno-0-658" name="__codelineno-0-658"></a><span class="sd">        two_dataframes (bool):</span>
<a id="__codelineno-0-659" name="__codelineno-0-659"></a><span class="sd">        sample_list (list):</span>
<a id="__codelineno-0-660" name="__codelineno-0-660"></a><span class="sd">        sample_list2 (list):</span>
<a id="__codelineno-0-661" name="__codelineno-0-661"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-662" name="__codelineno-0-662"></a>    <span class="c1"># output:</span>
<a id="__codelineno-0-663" name="__codelineno-0-663"></a>    <span class="c1">### clonoset_lists</span>
<a id="__codelineno-0-664" name="__codelineno-0-664"></a>
<a id="__codelineno-0-665" name="__codelineno-0-665"></a>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df</span><span class="p">):</span>
<a id="__codelineno-0-666" name="__codelineno-0-666"></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input clonosets in DataFrame have non-unique sample_id&#39;s&quot;</span><span class="p">)</span>
<a id="__codelineno-0-667" name="__codelineno-0-667"></a>    <span class="n">clonosets_df_1</span> <span class="o">=</span> <span class="n">clonosets_df</span><span class="p">[[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;filename&quot;</span><span class="p">]]</span>
<a id="__codelineno-0-668" name="__codelineno-0-668"></a>    <span class="n">two_dataframes</span> <span class="o">=</span> <span class="kc">False</span>
<a id="__codelineno-0-669" name="__codelineno-0-669"></a>    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<a id="__codelineno-0-670" name="__codelineno-0-670"></a>        <span class="n">two_dataframes</span> <span class="o">=</span> <span class="kc">True</span>
<a id="__codelineno-0-671" name="__codelineno-0-671"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="p">):</span>
<a id="__codelineno-0-672" name="__codelineno-0-672"></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Input clonosets in DataFrame2 have non-unique sample_id&#39;s&quot;</span><span class="p">)</span>
<a id="__codelineno-0-673" name="__codelineno-0-673"></a>        <span class="n">clonosets_df_2</span> <span class="o">=</span> <span class="n">clonosets_df2</span><span class="p">[[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;filename&quot;</span><span class="p">]]</span>
<a id="__codelineno-0-674" name="__codelineno-0-674"></a>        <span class="n">intersecting_sample_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">clonosets_df2</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">clonosets_df</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<a id="__codelineno-0-675" name="__codelineno-0-675"></a>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">intersecting_sample_ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">cl_filter2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-676" name="__codelineno-0-676"></a>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;WARNING! Some samples have the same sample_id in two sample_df&#39;s. The second filter will be applied to common samples&quot;</span><span class="p">)</span>
<a id="__codelineno-0-677" name="__codelineno-0-677"></a>
<a id="__codelineno-0-678" name="__codelineno-0-678"></a>
<a id="__codelineno-0-679" name="__codelineno-0-679"></a>    <span class="c1"># converting clonosets to compact lists of clonotypes separated by CDR3 lengths to dictionary based on overlap type and count/freq/umi</span>
<a id="__codelineno-0-680" name="__codelineno-0-680"></a>    <span class="n">clonoset_lists</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df_1</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">,</span>
<a id="__codelineno-0-681" name="__codelineno-0-681"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="n">strict</span><span class="p">)</span>
<a id="__codelineno-0-682" name="__codelineno-0-682"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-683" name="__codelineno-0-683"></a>        <span class="k">if</span> <span class="n">cl_filter2</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<a id="__codelineno-0-684" name="__codelineno-0-684"></a>            <span class="n">cl_filter2</span> <span class="o">=</span> <span class="n">cl_filter</span>
<a id="__codelineno-0-685" name="__codelineno-0-685"></a>        <span class="n">clonoset_lists_2</span> <span class="o">=</span> <span class="n">convert_clonosets_to_compact_dicts</span><span class="p">(</span><span class="n">clonosets_df_2</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter2</span><span class="p">,</span>
<a id="__codelineno-0-686" name="__codelineno-0-686"></a>                                                        <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="n">by_freq</span><span class="p">,</span> <span class="n">strict</span><span class="o">=</span><span class="n">strict</span><span class="p">)</span>
<a id="__codelineno-0-687" name="__codelineno-0-687"></a>        <span class="n">clonoset_lists</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">clonoset_lists_2</span><span class="p">)</span>
<a id="__codelineno-0-688" name="__codelineno-0-688"></a>
<a id="__codelineno-0-689" name="__codelineno-0-689"></a>    <span class="n">samples_total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">clonosets_df_1</span><span class="p">)</span>
<a id="__codelineno-0-690" name="__codelineno-0-690"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-691" name="__codelineno-0-691"></a>        <span class="n">samples_total</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">clonosets_df_1</span><span class="p">,</span> <span class="n">clonosets_df_2</span><span class="p">]))</span>
<a id="__codelineno-0-692" name="__codelineno-0-692"></a>
<a id="__codelineno-0-693" name="__codelineno-0-693"></a>    <span class="n">sample_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">clonosets_df_1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;sample_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sample_id</span><span class="p">)</span>
<a id="__codelineno-0-694" name="__codelineno-0-694"></a>    <span class="n">sample_list2</span> <span class="o">=</span> <span class="kc">None</span>
<a id="__codelineno-0-695" name="__codelineno-0-695"></a>    <span class="k">if</span> <span class="n">two_dataframes</span><span class="p">:</span>
<a id="__codelineno-0-696" name="__codelineno-0-696"></a>        <span class="n">sample_list2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">clonosets_df_2</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;sample_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sample_id</span><span class="p">)</span>
<a id="__codelineno-0-697" name="__codelineno-0-697"></a>
<a id="__codelineno-0-698" name="__codelineno-0-698"></a>    <span class="k">return</span> <span class="n">clonoset_lists</span><span class="p">,</span> <span class="n">samples_total</span><span class="p">,</span> <span class="n">two_dataframes</span><span class="p">,</span> <span class="n">sample_list</span><span class="p">,</span> <span class="n">sample_list2</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="intersections.tcrnet" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">tcrnet</span><span class="p">(</span><span class="n">clonosets_df_exp</span><span class="p">,</span> <span class="n">clonosets_df_control</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter_c</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s1">&#39;aaVJ&#39;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>This is an implementation of TCRnet (TCR neighbour enrichment test) algorithm.  It identifies similar clonotypes for the experimental dataset and the control one based on sequence 
similarity (allowing up to <code>mismatches</code> differences).  <br />
Args:
    clonosets_df_exp (pd.DataFrame): a DataFrame with experimental clonosets containing three columns - <code>sample_id</code> and <code>filename</code> columns,
        filename - full path to a clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
    clonosets_df_control (pd.DataFrame): a DataFrame with control clonosets containing three columns - <code>sample_id</code> and <code>filename</code> columns,
        filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format
    cl_filter (repseq.clone_filter.Filter): A filter applied to the experimental dataset before processing
    cl_filter_c (repseq.clone_filter.Filter): A filter applied to the control dataset before processing
    overlap_type (str): possible values are <code>aa</code>, <code>aaV</code>, <code>aaVJ</code>, <code>nt</code>, <code>ntV</code>, <code>ntVJ</code>. aa/nt define which CDR3 sequence
        to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments
        to decide if clonotypes are equal
    mismatches (int): Max number of single-letter mismatches in clonotype sequences 
        for them to be treated similar, i.e. hamming distance.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dataframe with following columns: <code>fold</code>, <code>p_value_b</code>, <code>p_value_p</code>, <code>p_value_b_adj</code>, <code>p_value_p_adj</code>, <code>log10_b_adj</code>, <code>log10_p_adj</code>, <code>log2_fc</code>. <code>p</code> in <code>p_value</code> </p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>stands for <code>poisson</code>, <code>b</code> for <code>binomial</code>, <code>adj</code> for multiple testing correction, and <code>log2_fc</code>for log2 fold change</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>repseq/intersections.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-0-270">270</a></span>
<span class="normal"><a href="#__codelineno-0-271">271</a></span>
<span class="normal"><a href="#__codelineno-0-272">272</a></span>
<span class="normal"><a href="#__codelineno-0-273">273</a></span>
<span class="normal"><a href="#__codelineno-0-274">274</a></span>
<span class="normal"><a href="#__codelineno-0-275">275</a></span>
<span class="normal"><a href="#__codelineno-0-276">276</a></span>
<span class="normal"><a href="#__codelineno-0-277">277</a></span>
<span class="normal"><a href="#__codelineno-0-278">278</a></span>
<span class="normal"><a href="#__codelineno-0-279">279</a></span>
<span class="normal"><a href="#__codelineno-0-280">280</a></span>
<span class="normal"><a href="#__codelineno-0-281">281</a></span>
<span class="normal"><a href="#__codelineno-0-282">282</a></span>
<span class="normal"><a href="#__codelineno-0-283">283</a></span>
<span class="normal"><a href="#__codelineno-0-284">284</a></span>
<span class="normal"><a href="#__codelineno-0-285">285</a></span>
<span class="normal"><a href="#__codelineno-0-286">286</a></span>
<span class="normal"><a href="#__codelineno-0-287">287</a></span>
<span class="normal"><a href="#__codelineno-0-288">288</a></span>
<span class="normal"><a href="#__codelineno-0-289">289</a></span>
<span class="normal"><a href="#__codelineno-0-290">290</a></span>
<span class="normal"><a href="#__codelineno-0-291">291</a></span>
<span class="normal"><a href="#__codelineno-0-292">292</a></span>
<span class="normal"><a href="#__codelineno-0-293">293</a></span>
<span class="normal"><a href="#__codelineno-0-294">294</a></span>
<span class="normal"><a href="#__codelineno-0-295">295</a></span>
<span class="normal"><a href="#__codelineno-0-296">296</a></span>
<span class="normal"><a href="#__codelineno-0-297">297</a></span>
<span class="normal"><a href="#__codelineno-0-298">298</a></span>
<span class="normal"><a href="#__codelineno-0-299">299</a></span>
<span class="normal"><a href="#__codelineno-0-300">300</a></span>
<span class="normal"><a href="#__codelineno-0-301">301</a></span>
<span class="normal"><a href="#__codelineno-0-302">302</a></span>
<span class="normal"><a href="#__codelineno-0-303">303</a></span>
<span class="normal"><a href="#__codelineno-0-304">304</a></span>
<span class="normal"><a href="#__codelineno-0-305">305</a></span>
<span class="normal"><a href="#__codelineno-0-306">306</a></span>
<span class="normal"><a href="#__codelineno-0-307">307</a></span>
<span class="normal"><a href="#__codelineno-0-308">308</a></span>
<span class="normal"><a href="#__codelineno-0-309">309</a></span>
<span class="normal"><a href="#__codelineno-0-310">310</a></span>
<span class="normal"><a href="#__codelineno-0-311">311</a></span>
<span class="normal"><a href="#__codelineno-0-312">312</a></span>
<span class="normal"><a href="#__codelineno-0-313">313</a></span>
<span class="normal"><a href="#__codelineno-0-314">314</a></span>
<span class="normal"><a href="#__codelineno-0-315">315</a></span>
<span class="normal"><a href="#__codelineno-0-316">316</a></span>
<span class="normal"><a href="#__codelineno-0-317">317</a></span>
<span class="normal"><a href="#__codelineno-0-318">318</a></span>
<span class="normal"><a href="#__codelineno-0-319">319</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-0-270" name="__codelineno-0-270"></a><span class="k">def</span><span class="w"> </span><span class="nf">tcrnet</span><span class="p">(</span><span class="n">clonosets_df_exp</span><span class="p">,</span> <span class="n">clonosets_df_control</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cl_filter_c</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="s2">&quot;aaVJ&quot;</span><span class="p">,</span> <span class="n">mismatches</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<a id="__codelineno-0-271" name="__codelineno-0-271"></a>
<a id="__codelineno-0-272" name="__codelineno-0-272"></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<a id="__codelineno-0-273" name="__codelineno-0-273"></a><span class="sd">    This is an implementation of TCRnet (TCR neighbour enrichment test) algorithm.  It identifies similar clonotypes for the experimental dataset and the control one based on sequence </span>
<a id="__codelineno-0-274" name="__codelineno-0-274"></a><span class="sd">    similarity (allowing up to `mismatches` differences).    </span>
<a id="__codelineno-0-275" name="__codelineno-0-275"></a><span class="sd">    Args:</span>
<a id="__codelineno-0-276" name="__codelineno-0-276"></a><span class="sd">        clonosets_df_exp (pd.DataFrame): a DataFrame with experimental clonosets containing three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-277" name="__codelineno-0-277"></a><span class="sd">            filename - full path to a clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-278" name="__codelineno-0-278"></a><span class="sd">        clonosets_df_control (pd.DataFrame): a DataFrame with control clonosets containing three columns - `sample_id` and `filename` columns,</span>
<a id="__codelineno-0-279" name="__codelineno-0-279"></a><span class="sd">            filename - full path to clonoset file. Clonoset file may be of MiXCR3/MiXCR4 or VDJtools format</span>
<a id="__codelineno-0-280" name="__codelineno-0-280"></a><span class="sd">        cl_filter (repseq.clone_filter.Filter): A filter applied to the experimental dataset before processing</span>
<a id="__codelineno-0-281" name="__codelineno-0-281"></a><span class="sd">        cl_filter_c (repseq.clone_filter.Filter): A filter applied to the control dataset before processing</span>
<a id="__codelineno-0-282" name="__codelineno-0-282"></a><span class="sd">        overlap_type (str): possible values are `aa`, `aaV`, `aaVJ`, `nt`, `ntV`, `ntVJ`. aa/nt define which CDR3 sequence</span>
<a id="__codelineno-0-283" name="__codelineno-0-283"></a><span class="sd">            to use (amino acid or nucleotide). V/J in the overlap_type define whether to check V or J segments</span>
<a id="__codelineno-0-284" name="__codelineno-0-284"></a><span class="sd">            to decide if clonotypes are equal</span>
<a id="__codelineno-0-285" name="__codelineno-0-285"></a><span class="sd">        mismatches (int): Max number of single-letter mismatches in clonotype sequences </span>
<a id="__codelineno-0-286" name="__codelineno-0-286"></a><span class="sd">            for them to be treated similar, i.e. hamming distance.</span>
<a id="__codelineno-0-287" name="__codelineno-0-287"></a>
<a id="__codelineno-0-288" name="__codelineno-0-288"></a><span class="sd">    Returns:</span>
<a id="__codelineno-0-289" name="__codelineno-0-289"></a><span class="sd">        df (pd.DataFrame): dataframe with following columns: `fold`, `p_value_b`, `p_value_p`, `p_value_b_adj`, `p_value_p_adj`, `log10_b_adj`, `log10_p_adj`, `log2_fc`. `p` in `p_value` </span>
<a id="__codelineno-0-290" name="__codelineno-0-290"></a><span class="sd">        stands for `poisson`, `b` for `binomial`, `adj` for multiple testing correction, and `log2_fc`for log2 fold change </span>
<a id="__codelineno-0-291" name="__codelineno-0-291"></a><span class="sd">    &quot;&quot;&quot;</span>
<a id="__codelineno-0-292" name="__codelineno-0-292"></a>
<a id="__codelineno-0-293" name="__codelineno-0-293"></a>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Running TCRnet neighbour count</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;-&quot;</span><span class="o">*</span><span class="mi">50</span><span class="p">)</span>
<a id="__codelineno-0-294" name="__codelineno-0-294"></a>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overlap type: </span><span class="si">{</span><span class="n">overlap_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<a id="__codelineno-0-295" name="__codelineno-0-295"></a>
<a id="__codelineno-0-296" name="__codelineno-0-296"></a>    <span class="n">clonoset_exp</span> <span class="o">=</span> <span class="n">pool_clonotypes_from_clonosets_df</span><span class="p">(</span><span class="n">clonosets_df_exp</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter</span><span class="p">)</span>
<a id="__codelineno-0-297" name="__codelineno-0-297"></a>    <span class="n">clonoset_exp_dict</span> <span class="o">=</span> <span class="n">prepare_clonoset_for_intersection</span><span class="p">(</span><span class="n">clonoset_exp</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">len_vj_format</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-298" name="__codelineno-0-298"></a>
<a id="__codelineno-0-299" name="__codelineno-0-299"></a>    <span class="n">unique_clonotypes</span> <span class="o">=</span> <span class="p">[(</span><span class="n">seq_count</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="o">*</span><span class="n">len_vj</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span> <span class="k">for</span> <span class="n">len_vj</span><span class="p">,</span> <span class="n">seq_counts</span> <span class="ow">in</span> <span class="n">clonoset_exp_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">for</span> <span class="n">seq_count</span> <span class="ow">in</span> <span class="n">seq_counts</span><span class="p">]</span>
<a id="__codelineno-0-300" name="__codelineno-0-300"></a>
<a id="__codelineno-0-301" name="__codelineno-0-301"></a>    <span class="n">clonoset_control</span> <span class="o">=</span> <span class="n">pool_clonotypes_from_clonosets_df</span><span class="p">(</span><span class="n">clonosets_df_control</span><span class="p">,</span> <span class="n">cl_filter</span><span class="o">=</span><span class="n">cl_filter_c</span><span class="p">)</span>
<a id="__codelineno-0-302" name="__codelineno-0-302"></a>    <span class="n">clonoset_control_dict</span> <span class="o">=</span> <span class="n">prepare_clonoset_for_intersection</span><span class="p">(</span><span class="n">clonoset_control</span><span class="p">,</span> <span class="n">overlap_type</span><span class="o">=</span><span class="n">overlap_type</span><span class="p">,</span> <span class="n">by_freq</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">len_vj_format</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<a id="__codelineno-0-303" name="__codelineno-0-303"></a>
<a id="__codelineno-0-304" name="__codelineno-0-304"></a>
<a id="__codelineno-0-305" name="__codelineno-0-305"></a>    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
<a id="__codelineno-0-306" name="__codelineno-0-306"></a>    <span class="n">chunks</span> <span class="o">=</span> <span class="mi">40</span>
<a id="__codelineno-0-307" name="__codelineno-0-307"></a>    <span class="n">chunk_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_clonotypes</span><span class="p">)</span><span class="o">//</span><span class="n">chunks</span><span class="o">+</span><span class="mi">1</span>
<a id="__codelineno-0-308" name="__codelineno-0-308"></a>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">chunks</span><span class="p">):</span>
<a id="__codelineno-0-309" name="__codelineno-0-309"></a>        <span class="n">first</span> <span class="o">=</span> <span class="n">i</span><span class="o">*</span><span class="n">chunk_size</span>
<a id="__codelineno-0-310" name="__codelineno-0-310"></a>        <span class="n">last</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">chunk_size</span>
<a id="__codelineno-0-311" name="__codelineno-0-311"></a>        <span class="n">task</span> <span class="o">=</span> <span class="p">(</span><span class="n">unique_clonotypes</span><span class="p">[</span><span class="n">first</span><span class="p">:</span><span class="n">last</span><span class="p">],</span> <span class="n">clonoset_exp_dict</span><span class="p">,</span> <span class="n">clonoset_control_dict</span><span class="p">,</span> <span class="n">mismatches</span><span class="p">)</span>
<a id="__codelineno-0-312" name="__codelineno-0-312"></a>        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">task</span><span class="p">)</span>
<a id="__codelineno-0-313" name="__codelineno-0-313"></a>
<a id="__codelineno-0-314" name="__codelineno-0-314"></a>    <span class="n">results</span> <span class="o">=</span> <span class="n">run_parallel_calculation</span><span class="p">(</span><span class="n">tcrnet_mp</span><span class="p">,</span> <span class="n">tasks</span><span class="p">,</span> <span class="s2">&quot;Calc neighbours (TCRnet)&quot;</span><span class="p">,</span> <span class="n">object_name</span><span class="o">=</span><span class="s2">&quot;parts&quot;</span><span class="p">)</span>
<a id="__codelineno-0-315" name="__codelineno-0-315"></a>    <span class="n">results</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">results</span><span class="p">))</span> <span class="c1"># unpack results from several workers</span>
<a id="__codelineno-0-316" name="__codelineno-0-316"></a>
<a id="__codelineno-0-317" name="__codelineno-0-317"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;clone&quot;</span><span class="p">,</span> <span class="s2">&quot;count_exp&quot;</span><span class="p">,</span> <span class="s2">&quot;count_control&quot;</span><span class="p">,</span> <span class="s2">&quot;group_count_exp&quot;</span><span class="p">,</span> <span class="s2">&quot;group_count_control&quot;</span><span class="p">])</span>
<a id="__codelineno-0-318" name="__codelineno-0-318"></a>    <span class="n">df</span> <span class="o">=</span> <span class="n">tcrnet_stats_calc</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<a id="__codelineno-0-319" name="__codelineno-0-319"></a>    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy; 2023 <a href="https://github.com/mmjmike" target="_blank" rel="noopener">Mikhail Myshkin</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/mmjmike" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M202.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M496 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.path", "toc.integrate", "navigation.top", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>